# Genotype-Wise mRNA-Imputation
```{r setup-impute, echo = FALSE}
pacman::p_load("sspredr", "tidyverse", "caret", "e1071", "viridis",
               "forcats", "knitr", "rmarkdown", "purrr", "rprojroot",
               "methods", "parallel")
knitr::knit_hooks$set(inline = function(x) {
  prettyNum(x, big.mark = ",", small.mark = ".", digits = 2)
})
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(fig.width = 14, fig.height = 10, 
                      warning = FALSE, message = FALSE, echo = FALSE)
```


```{r Predictor-Data, cache = TRUE}
# Common genotypes
genos <- readRDS("./data/processed/common_genotypes.RDS")
mrna_inbreds <- genos %>%
  filter(Data_Type == "mrna") %>%
  split(.$Pool) %>%
  map("G")

# Remove the 'Group' variable after splitting Dent and Flint data.
rm_grp <- function(x) {
  x$Group <- NULL
  x
}

# mRNA
mrna <- readRDS("./data/processed/subset_mrna_blues.RDS")[["100%"]]
mrna <- t(mrna)
mrna <- mrna[grep("Exp|Sigma", x = rownames(mrna), invert = TRUE), ]
mrna_lst <- mrna %>%
  as_data_frame() %>%
  mutate(G = rownames(mrna)) %>%
  mutate(Group = ifelse(G %in% mrna_inbreds$Dent, yes = "Dent", no = "Flint"),
         Group = as.factor(Group)) %>%
  as.data.frame() %>%
  column_to_rownames(var = "G") %>%
  split(.$Group) %>%
  map(rm_grp) %>%
  map(~as.matrix(.))

# SNPs
snp <- readRDS("./data/processed/snp_mat.RDS")
snp_lst <- snp %>%
  as_data_frame() %>%
  mutate(G = rownames(snp)) %>%
  filter(G %in% (mrna_inbreds %>% flatten_chr())) %>%
  mutate(Group = ifelse(G %in% mrna_inbreds$Dent, yes = "Dent", no = "Flint"),
         Group = as.factor(Group)) %>%
  as.data.frame() %>%
  column_to_rownames(var = "G") %>%
  split(.$Group) %>%
  map(rm_grp) %>%
  map(~as.matrix(.))

# Pedigree
ped <- readRDS("./data/processed/ped-datafull-GTP.RDS")
ped_lst <- ped %>%
  mutate(G = rownames(ped)) %>%
  filter(G %in% (mrna_inbreds %>% flatten_chr())) %>%
  mutate(Group = ifelse(G %in% mrna_inbreds$Dent, yes = "Dent", no = "Flint"),
         Group = as.factor(Group)) %>%
  as.data.frame() %>%
  column_to_rownames(var = "G") %>%
  split(.$Group) %>%
  map(rm_grp) %>%
  map(as.matrix) %>%
  map(function(x) x[, match(rownames(x), colnames(x))]) %>%
  map(function(x) x * 2) %>%
  map(function(x) {
    x[is.na(x)] <- 0
    x
  })
```



```{r Impute-mRNA-Function}
## Function for the imputation of a single mRNA
impute_mrna_genotype <- function(x, y) {
  geno <- rownames(x)
  # Names of genotypes for which transcriptomic records exist
  nm2 <- rownames(y)
  # Names of genotypes for which transcriptomic records are missing.
  nm1 <- setdiff(rownames(x), nm2)
  # Hybrid parents not in y.
  geno1 <- geno[geno %in% nm1]
  # Hybrid parents in y.
  geno2 <- geno[geno %in% nm2]
  x <- x[match(unique(geno), rownames(x)), ]
  # Design matrix mapping GCA effects from genotypes with transcriptomic
  # records to y2.
  Z2 <- Matrix::sparse.model.matrix(~-1 + factor(geno2),
                                    drop.unused.levels = FALSE)
  colnames(Z2) <- gsub("factor\\(geno2\\)", replacement = "", x = colnames(Z2))
  Z2 <- Z2[, match(nm2, colnames(Z2))]
  rownames(Z2) <- geno2
  # Hybrid parents not in y.
  geno1 <- geno[geno %in% nm1]
  # Hybrid parents in y.
  geno2 <- geno[geno %in% nm2]
  x <- x[match(unique(geno), rownames(x)), ]
  y <- y[nm2, ]
  M2 <- y[, matrixStats::colVars(y) != 0]
  x <- x[, matrixStats::colVars(x) != 0]
  A <- build_kernel(M = x, lambda = 0.01, algorithm = "RadenII")
  A11 <- A[nm1, nm1, drop = FALSE]
  A12 <- A[nm1, nm2, drop = FALSE]
  A21 <- A[nm2, nm1, drop = FALSE]
  A22 <- A[nm2, nm2, drop = FALSE]
  Ainv <- solve(A)
  dimnames(Ainv) <- dimnames(A)
  A_up11 <- Ainv[nm1, nm1]
  A_up12 <- Ainv[nm1, nm2]
  # Eq.21
  M1 <- A12 %*% solve(A22) %*% M2
  J2 <- matrix(-1, nrow = ncol(A12), ncol = 1)
  # Eq.22
  J1 <- A12 %*% solve(A22) %*% J2
  # Eq.10
  epsilon <- t(chol(solve(Ainv[nm1, nm1])))
  # Eq.20
  W2 <- Z2 %*% M2
  w1 <- A12 %*% solve(A22) %*% Z2 %*% M2
  as.matrix(w1)
}

# Shuffled version of the imputation function above.
shuffled_imputation <- function(x, y) {
  cor_lst <- lapply(seq(from = 1, to = nrow(y) - 2), FUN = function(i) {
    idx <- sample(nrow(y), size = i, replace = FALSE)
    z <- y[-idx, ]
    imp_mrna <- impute_mrna_genotype(x = x, y = z)
    orig_mrna <- y[match(rownames(imp_mrna), rownames(y)), , drop = FALSE]
    stopifnot(identical(rownames(imp_mrna), rownames(orig_mrna)))
    cor(c(imp_mrna), c(orig_mrna))
  })
  cor_lst %>%
    flatten_dbl() %>%
    tbl_df() %>%
    rownames_to_column(var = "Number_NA_Geno") %>%
    rename(Value = value)
}
```


## Correlations between vectors of original and imputed mRNAs.
### General idea:
*    Impute transcriptomic data for different numbers of genotypes (from 1 to
     number of genotypes with transcriptomic data).
*    For each number of genotypes without transcriptomic data, randomly sample 
     1000 sets of genotypes, which will be imputed.
     This will allow us to aggregate and summarize the correlation between 
     observed and imputed mRNAs with high confidence.


The computations were run using multiple cores with 
[this script](../analysis/impute_single_mrna.R).
```{r Shuffle-Genotypes-Correlate-Imputation, eval = FALSE}
repl <- 1000
donors <- c("snp_lst", "ped_lst")
imp_lst <- lapply(donors, FUN = function(type) {
  par_lst <- vector(mode = "list", length = repl)
  names(par_lst) <- paste0("Replication_", seq_len(repl))
  par_lst[] <- mclapply(par_lst, FUN = function(iter, ...) {
    try(map2(.x = get(type), .y = mrna_lst, .f = shuffled_imputation))
  }, mc.cores = use_cores)
  par_lst
})
names(imp_lst) <- donors
```



```{r Plot-Correlation-Imputed-vs-NonImputed, cache = TRUE, fig.width = 12, fig.height = 12, fig.cap = "Pearson correlation coefficients between original and imputed mRNA genotypes for two heterotic groups and two sources of information (i.e. genomic and pedigree), depending on the number of genotypes without mRNA records. Genotypes without mRNA records were declared as such at random."}
par_lst <- readRDS("./data/derived/impute_single_mrna.RDS")
par_lst %>%
  map(~keep(., is_list)) %>%
  at_depth(.depth = 2, ~bind_rows(., .id = "Group")) %>%
  map(~bind_rows(., .id = "Replication")) %>%
  bind_rows(.id = "Data_Type") %>%
  mutate(Data_Type = fct_recode(Data_Type,
                                Genomic = "snp_lst",
                                Pedigree = "ped_lst")) %>%
  mutate(Number_NA_Geno = as.factor(Number_NA_Geno)) %>%
  mutate(Number_NA_Geno = fct_inorder(Number_NA_Geno)) %>%
  ggplot(aes(x = Number_NA_Geno, y = Value)) +
  geom_boxplot() +
  facet_wrap(Data_Type ~ Group, scales = "free") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1)) +
  xlab("Number of Genotypes Without mRNA-Data") +
  ylab("Correlation Between Original and Imputed mRNA BLUEs")
```
