---
title: Single-step Hybrid Predition
author: Matthias Westhues, Claas Heuer
date: "`r Sys.Date()`"
knit: "bookdown::render_book"
output:
  bookdown::html_document2:
    toc: true
    toc_float: true
    theme: readable
    highlight: tango
documentclass: book
bibliography: ["library.bib", "cheuer.bib"]
biblio-style: apalike
link-citations: yes
---

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>


# Preface {-}
This report covers an extension of the methodology of single-step prediction 
&mdash; which was developed for the imputation of SNP data from pedigree 
information &mdash; to the imputation of quantitative predictor variables using
genomic information.

Here, we are working with a "full" data set &mdash; referring to the maximum 
number of hybrids for which agronomic and genomic, but not necessarily 
transcriptomic data are available &mdash; and a "reduced" data set with less 
hybrids but all of their parental inbred lines having complete genomic and 
transcriptomic records.

All our predictive abilities were evaluated using a leave-one-out cross 
validation (LOOCV) routine and results are displayed for T0-hybrids only.
Note, that T0-hybrids are defined as hybrids having zero parents that have been
previously tested in other hybrid combinations.

The companion R-package for this analysis is [sspredr](https://github.com/mwesthues/sspredr).
The development version can be installed from Github:
```{r Install_sspred, eval = FALSE}
devtools::install_github("mwesthues/sspredr")
```


## Objectives
1.    Assess whether transcriptomic data with missing records for some genotypes         can be imputed successfully based on genomic information to improve 
      prediction of unseen hybrids.
2.    Explore the utility of various machine learning algorithms for the 
      prediction of unseen hybrids.^[This is motivated by the fact that the 
      success of hybrid prediction is a strongly convex function of predictive 
      ability.]


## Overview
Chapter \@ref(introduction) gives a summary of the relevant literature.

Chapter \@ref(data-overview) provides a brief summary of the available data 
without delving into their properties.

Chapter \@ref(agronomic-data) gives an overview of the distribution of the 
agronomic traits as well as correlations among them, separately for the full
set of hybrids as well as the reduced set of hybrids.

Chapter \@ref(transcriptomic-data) explores the distribution and correlations
of/among transcriptomic predictors for "raw" and transformed mRNA BLUEs.

Chapter \@ref(genotype-wise-mrna-imputation) examines the correlation between
the original and imputed transcriptomic BLUEs.

Chapter \@ref(predictor-subsetting) introduces the different predictor sets that
were used for predictions.

Chapter \@ref(loocv-prediction) deals with the results of hybrid prediction for 
the reduced set of data and a comparison of predictive abilities with the full 
data set.
