# Introduction
The integration of different sources of information for the prediction of 
breeding or production values is one of the major challenges in current genomic 
prediction research.
In the early years of genomic prediction, a 2-step procedure was the most 
intuitive choice for establishing a new selection scheme.
The first consists of conventionally estimating breeding values based on 
pedigree relationships and therefore the phenotypes of relatives.
Individuals with accurate breeding values out of the first step (i.e. high 
number of offspring), enter the second step in which SNP genotypes are regressed 
against the conventional breeding values.
Those markers can then be used to predict genomic breeding values for solely 
genotyped and potentially very young individuals.
Considering the two sources of information &mdash; pedigree and high density SNP 
genotypes &mdash; the process can be described as follows:


### First Step
Denote $\mathbf{y}$ as the vector of measured phenotypes for a normally 
distributed trait for all individuals available in the population.
The relationship &mdash; and therefore the strucure of the additive genetic 
covariance &mdash; across individuals is based on the numerator relationship 
matrix $\mathbf{A}$.
In order to obtain breeding values for any individual in the population, we 
consider the following model:

$$
\begin{equation}
	\mathbf{y} = \mathbf{Xb} + \mathbf{Za} + \mathbf{e},
	\label{eq:animal_model}
\end{equation}
$$

where $\mathbf{X}$ is a design matrix for fixed effects, $\mathbf{b}$ is a 
vector of fixed effects, $\mathbf{Z}$ is the design matrix for the random animal 
effects, which associate observations in $\mathbf{y}$ with elements in 
$\mathbf{A}$, $\mathbf{a}$ is the vector of random animal effects (=breeding 
values) and $\mathbf{e}$ is a vector of iid residual errors with expectation 
$0$ and variance $\sigma^2_e$.
The distribution of the random vector $\mathbf{u}$ is 
$MVN(0, \mathbf{A}\sigma^2_a)$.



### Second Step
In the second step, a subset of the individuals for which conventional breeding 
values have been estimated, is selected and enters the reference population for 
training a genomic prediction model.
That selection requires that the individual is genotyped and has 
$\text{PEV}(u_{\text{candidate}}) \leq \text{threshold}$.
As the breeding values from step one enter the genomic prediction model as 
phenotypes, one has to be careful about the residual error distribution of
such a model.
For instance, if the range of the prediction error variances is very broad in 
the reference population, the residual errors will not be identically 
distributed anymore, e.g. they should be weighted differently or estimated 
individually.
One popular approach to account for that problem is *Deregressing* the breeding 
values and weighting the residuals according to the prediction error variances 
[@Garrick2009].
Genomic breeding values for any animal in the population can then be obtained 
by:

$$
\begin{equation}
	\label{eq:gebv}
	\mathbf{\hat{g}} =  (\mathbf{M} - 2\mathbf{p})\mathbf{\hat{u}},
\end{equation}
$$

where $\mathbf{p}$ is a row-vector of allele frequencies for all the markers in 
$\mathbf{M}$ in the base population.
The prediction error variance of the vector $\mathbf{\hat{g}}$ is usually 
globally obtained using cross validation. 
Alternatively they could be obtained individually by the posterior distribution 
of $(\mathbf{M} - 2\mathbf{p})\mathbf{\hat{u}}$ when a Bayesian model is used.


## Single Step Transciptomic prediction of Breeding Values
The breeding values in a mRNA prediction model are in general given by:
$$
\begin{equation}
	\label{eq:mrnaebv}
	\mathbf{\hat{g}} = \mathbf{W}\boldsymbol{\hat{\alpha}},
\end{equation}
$$

where $\boldsymbol{\hat{\alpha}}$ is the solution to a ridge regression model 
equaivalent to the GBLUP model but using mRNA covariates instead of 
marker genotypes.

Consider now the situation in which only a subset of the population has 
transcriptomic information but all have SNP genotypes.
We can take a similar route as in @fernando_class_2014 and impute missing
mRNA data.
The difference here is that &mdash; instead of predicting SNP covariates using 
pedigree relationship &mdash; we impute MRNA covariates using SNP or genomic 
relationships.
Let the subscript $1$ denote individuals for which SNP genotypes but no mRNA 
data are available.