\documentclass[]{elsarticle} %review=doublespace preprint=single 5p=2 column
%%% Begin My package additions %%%%%%%%%%%%%%%%%%%
\usepackage[hyphens]{url}
\usepackage{lineno} % add
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}

\bibliographystyle{elsarticle-harv}
\biboptions{sort&compress} % For natbib
\usepackage{graphicx}
\usepackage{booktabs} % book-quality tables
%% Redefines the elsarticle footer
%\makeatletter
%\def\ps@pprintTitle{%
% \let\@oddhead\@empty
% \let\@evenhead\@empty
% \def\@oddfoot{\it \hfill\today}%
% \let\@evenfoot\@oddfoot}
%\makeatother

% A modified page layout
\textwidth 6.75in
\oddsidemargin -0.15in
\evensidemargin -0.15in
\textheight 9in
\topmargin -0.5in
%%%%%%%%%%%%%%%% end my additions to header

\usepackage[T1]{fontenc}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \usepackage{fontspec}
  \ifxetex
    \usepackage{xltxtra,xunicode}
  \fi
  \defaultfontfeatures{Mapping=tex-text,Scale=MatchLowercase}
  \newcommand{\euro}{€}
\fi
% use microtype if available
\IfFileExists{microtype.sty}{\usepackage{microtype}}{}
\usepackage{longtable}
\usepackage{graphicx}
% We will generate all images so they have a width \maxwidth. This means
% that they will get their normal width if they fit onto the page, but
% are scaled down if they would overflow the margins.
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth
\else\Gin@nat@width\fi}
\makeatother
\let\Oldincludegraphics\includegraphics
\renewcommand{\includegraphics}[1]{\Oldincludegraphics[width=\maxwidth]{#1}}
\ifxetex
  \usepackage[setpagesize=false, % page size defined by xetex
              unicode=false, % unicode breaks when used with xetex
              xetex]{hyperref}
\else
  \usepackage[unicode=true]{hyperref}
\fi
\hypersetup{breaklinks=true,
            bookmarks=true,
            pdfauthor={},
            pdftitle={Omics-based Single Step Trait Prediction},
            colorlinks=true,
            urlcolor=blue,
            linkcolor=magenta,
            pdfborder={0 0 0}}
\urlstyle{same}  % don't use monospace font for urls
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\setcounter{secnumdepth}{5}
% Pandoc toggle for numbering sections (defaults to be off)
% Pandoc header


\usepackage[nomarkers]{endfloat}

\begin{document}
\begin{frontmatter}

  \title{Omics-based Single Step Trait Prediction}
    \author[University of Hohenheim]{Matthias Westhues}
   \ead{matthias.westhues@uni-hohenheim.de} 
  
    \author[University of Kiel]{Claas Heuer}
   \ead{cheuer@tierzucht.uni-kiel.de} 
  
    \author[University of Kiel]{Georg Thaller}
   \ead{thaller@tierzucht.uni-kiel.de} 
  
    \author[Iowa State University]{Rohan Fernando}
   \ead{rohan@iastate.edu} 
  
    \author[University of Hohenheim]{Albrecht E. Melchinger\corref{c1}}
   \ead{melchinger@uni-hohenheim.de} 
   \cortext[c1]{Corresponding Author}
      \address[University of Hohenheim]{Institute of Plant Breeding, Seed Science and Population Genetics,
Fruwirthstr. 21, 70599 Stuttgart, Germany}
    \address[University of Kiel]{Institut of Animal Breeding and Husbandry, Hermann-Rodewald-Straße 6,
24118 Kiel, Germany}
    \address[Iowa State University]{Department of Animal Science, 239 Kildee Hall, 50011 Ames, Iowa, U.S.A.}
  
  \begin{abstract}
  \(\dots\)
  \end{abstract}
  
 \end{frontmatter}

\section{Introduction}\label{introduction}

\(\dots\)

Our objectives were threefold: (1) Examine whether the single step
prediction framework can be transfered to hybrids, (2) explore the
utility of single step predition with a quantitative predictor and (3)
evaluate the impact of the particular genetic space covered by the
incomplete predictor on predictive ability.

\section{Materials and Methods}\label{materials-and-methods}

\subsection{Data}\label{data}

\subsubsection{Hybrid data}\label{hybrid-data}

The maize hybrid data set comprises 1,521 hybrids produced in 16
factorial mating designs between 142 Dent and 103 Flint parent lines
(Westhues et al. 2017). Best linear unbiased estimates (BLUEs) were
computed across all factorials and three or more agro-ecologically
diverse locations across Germany for six agronomically important traits.
All 245 parent lines have pedigree records back to the generation of
their grandparents or deeper (Westhues et al. 2017) and were genotyped
using the Illumina BeadChip MaizeSNP50 (Ganal et al. 2011). After
filtering for minor allele frequency (\(\geq 5\)\%), heterozygosity rate
(\(\leq 5\)\%) and call frequency (\(\geq 95\)\%), missing genotypes
were imputed using the Beagle software (B. L. Browning and Browning
2009). The final number of polymorphic marker loci was 7,013 for
parental Dent and 6,212 for parental Flint lines, respectively. Gene
expression data for seedlings seven days after sowing were obtained for
a subset of 60 parental Dent and 43 parental Flint lines through
two-color hybridizations using a custom 2K microarray (GPL22267)
(Westhues et al. 2017). After applying established normalization
procedures (G. K. Smyth and Speed 2003; Matthew E Ritchie et al. 2007),
gene expression BLUEs were computed for 1,323 transcripts in reference
to established protocols (G. K. Smyth and Speed 2003; Matthew E Ritchie
et al. 2007; Frisch et al. 2010) using the R-package \emph{limma} (M. E.
Ritchie et al. 2015).

\subsubsection{Diversity panel}\label{diversity-panel}

The maize inbred line data set originally comprised 513 maize lines
representing the global maize diversity and was reduced to the set of
tropical and subtropical lines (\(n = 211\)) given that it is the
largest of the four pre-classified subgroups (N. Yang et al. 2014). All
inbred lines were evaluated in five Chinese environments ranging from
\(18\) to \(30^{\circ}\)N and from \(102\) to \(110^{\circ}\)E (N. Yang
et al. 2014). Best linear unbiased predictors (BLUPs) were calculated
for 17 traits of which six were used for this study. MaizeSNP50
BeadChip-data were available for the entire set of lines from the
diversity panel. Additionally, RNA sequencing was performed for 368
lines on young seedlings 15 days after sowing. By exploting identify by
descent (IBD) between 49,728 SNPs on the BeadChip overlapping with the
RNA-seq data, 556,809 high quality SNPs could be inferred for these
lines (Fu et al. 2013; H. Li et al. 2013). For the remaining 145 maize
lines without RNA-seq data, high density markers were inferred via
projection of IBD regions onto the BeadChip, using this core set of
markers (N. Yang et al. 2014). The same SNP quality checks as for the
parental inbred lines of the hybrids were applied to the 211 inbred
lines from the tropical/subtropical subset, yielding 37,760 SNPs. Gene
expression data were normalized using a normal quantile transformation
to satisfy modelling assumptions (Fu et al. 2013). A subset of 28,850
annotated genes was available for 149 out of the 211 genotypes and was
kept for further analyses (H. Li et al. 2013).

\subsection{Population structure}\label{population-structure}

Principal component analyses (PCA) were run on scaled and centered
predictor matrices using the \emph{pca()} function from the R-package
\emph{LEA} (Frichot and François 2015). Estimates of ancestry
coefficients were estimated based on sparse nonnegative matrix
factorization algorithms (Frichot et al. 2014) using the \emph{snmf()}
function in \emph{LEA}. Algorithms were run for \(K=2, \dots 5\)
putative ancestral gene pools with 25 repetitions for each value of
\(K\). For each \(K\), only the repetition with the lowest cross-entropy
was kept for further analyses.

\subsection{Core Sampling}\label{core-sampling}

In order to evaluate the influence of the genetic constitution of the
set of genotypes that has only data on one out of two predictors, we
generated core samples using the R-package \emph{corehunter} (Thachuk et
al. 2009). These core samples were based on the 149 maize inbred lines
from the diversity panel which had gene expression as well as genomic
information. The size of the core set was varied in increments of ten
percentage points and ranged from 10\% to 90\% of all 149 inbred lines.
For assembling the various core sets, we first computed the Modifed
Rogers (MR) distance between each pair of genotypes and then selected a
predefined subset of genotypes that display maximum average distance
between each other. MR was computed using the \emph{rogers.dist()}
function from the R-pacakge \emph{poppr} (Kamvar, Brooks, and Grünwald
2015).

\subsection{Prediction}\label{prediction}

\subsubsection{Kernels}\label{kernels}

BLUP models are computationally attractive when the number of features
\(p\) exceeds the number of genotypes \(n\) and are equivalent to a
selection index when fixed effects have been accomodated by the
dependent variable (Mrode 2014). Depending on the data set, up to three
predictors were available for agronomic trait predictions, namely
pedigree data (P), genomic data (G) and transcriptompic data (T). The
corresponding feature matrix of the \(l\)-th group of inbred lines and
the \(o\)-th predictor (\(\mathbf{W}_{lo}\)) has dimensions
\(n \times p\) where \(n\) pertains to the number of genotypes in the
\(l\)-th group and \(p\) pertains to the number of features. For the
diversity panel maize lines, \(n\) corresponds to the number of
genotypes whereas, in the case of the hybrid data, \(n\) corresponds to
the number of parent lines in the corresponding heterotic group. In
generating kernels for each predictor, all features in
\(\mathbf{W}_{lo}\) were centered and standardized to unit variance.
Then, each kernel can be defined as

\begin{equation} \label{eq:GenomicRelationship}
  \mathbf{K}_{lo} = \frac{1}{W_{lo}} \mathbf{W}_{lo} \mathbf{W}_{lo}^{\top},
\end{equation}

where \(W_{lo}\) denotes the number of features (VanRaden 2008). In the
case of pedigree data (P), coancestry coefficients were used directly
for \(\mathbf{K}_{lo}\).

\subsubsection{Models}\label{models}

The universal model for breeding values was:

\begin{equation} \label{eq:KBLUPModel}
  \mathbf{y} = \mu + 
  \sum_{l=1}^{L} \mathbf{Z}_{lo} \mathbf{g}_{lo} +
  \mathbf{\epsilon},
\end{equation}

where \(\mathbf{y}\) is the vector of observed inbred line or hybrid
performance, respectively, \(\mu\) is the fixed model intercept,
\(\mathbf{Z}_{lo}\) is the corresponding design matrix associating the
random genotype effects (\emph{i.e.}, breeding values) of the lines in
the \(l\)-th inbred line group and the \(o\)-th predictor type with
\(\mathbf{y}\). Note that, in the case of hybrid data,
\(\mathbf{Z}_{lo}\) associates the random general combining ability
effects in the parental inbred lines (either Dent or Flint, as expressed
by \(l\)) with the vector of observed hybrid performance \(\mathbf{y}\).
The random effects (\(\mathbf{g}_{lo}\)) have expectation zero and
variance equal to \(\mathbf{K}_{lo} \sigma^{2}_{{g}^{lo}}\) and
\(\mathbf{I} \sigma^2_{\epsilon}\) for the residual error. All
prediction models were implemented using the R-package \emph{BGLR}
(Pérez and de Los Campos 2014).

\subsubsection{Single Step Prediction}\label{single-step-prediction}

The breeding values in a prediction model are in general given by:

\begin{equation} \label{eq:mrnaebv}
    \mathbf{\tilde{g}} = \mathbf{W}\boldsymbol{\hat{\alpha}},
\end{equation}

where \(\boldsymbol{\hat{\alpha}}\) is the solution to a ridge
regression model, which is equivalent to the BLUP model.

Consider now the situation in which one predictor is complete in the
sense that it contains information on the whole population whereas a
second predictor is incomplete and covers only a subset of the
population. We can take a similar route as in Fernando, Dekkers, and
Garrick (2014) and impute covariates of the incomplete predictor by
using covariates of the complete predictor. Let the subscript \(1\)
denote individuals covered only by the complete predictor whereas
individuals covered by both, the complete and the incomplete predictor,
are indicated by the subscript \(2\). Further, the covariates in
\(\mathbf{W}\) are centered. The vector \(\mathbf{g}_1\) can be written
as the sum of the conditional expectation given \(\mathbf{g}_2\) and a
residual error term:

\begin{align} \label{eq:mrna1}
    \mathbf{g_1} &= E(\mathbf{g}_1|\mathbf{g}_2) + \boldsymbol{\epsilon} \\
    &= \mathbf{K_{12}}\mathbf{K_{22}}^{-1}\mathbf{W_2}\boldsymbol{\hat{\alpha}} + (\mathbf{g_1} - \mathbf{K_{12}}\mathbf{K_{22}}^{-1}\mathbf{W_2}\boldsymbol{\hat{\alpha}}) \\
    &= \mathbf{\hat{g}}_1 + \boldsymbol{\epsilon}
\end{align}

The covariance matrix of \(\boldsymbol{\epsilon}\) is
\(\mathbf{K}_{11} - \mathbf{K}_{12}\mathbf{K}_{22}^{-1}\mathbf{K}_{21} = (\mathbf{K}^{11})^{-1}\)
(Legarra, Aguilar, and Misztal 2009).

The structure of the residual imputation/prediction error is known and
can therefore be modelled. The covariates not covered by the incomplete
predictor can be predicted using the expectation of a multivariate
normal random vector given correlated observations, which are related to
the Best Linear Predictor.

The general model for the single step procedure can then be written as

\begin{equation}
  \label{eq:single-step-model}
\mathbf{y} = \mathbf{Xb} + \mathbf{W} \boldsymbol{\alpha} + \mathbf{U}^{Complete} \boldsymbol{\epsilon}^{Complete} + \mathbf{e},
\end{equation}

with

\begin{equation}
 \label{eq:single-step-submatrices}
\mathbf{X} = 
\begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix},
 \mathbf{W} = 
\begin{bmatrix}
  Z_1\hat{\mathbf{W}_1} \\
  \mathbf{W}_2 
 \end{bmatrix},
 \mathbf{U}^{Complete} = 
\begin{bmatrix}
  Z_1 \\
  0 
 \end{bmatrix}
\end{equation}

Following the notation in Fernando, Dekkers, and Garrick (2014) the
phenotypes can be untangled like this:

\begin{align}
 \label{eq:entangled-augmented-single-step-model}
\begin{bmatrix}
  y_1 \\
  y_2 
 \end{bmatrix}
& =
 \begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix}
 \boldsymbol{\beta} + 
 \begin{bmatrix}
  Z_1 & 0 \\
  0 & Z_2 
 \end{bmatrix}
\begin{bmatrix}
  g_1 \\
  g_2 
 \end{bmatrix}
  + \mathbf{e} \\
    & = 
 \begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix}
 \boldsymbol{\beta} + 
 \begin{bmatrix}
  Z_1 & 0 \\
  0 & Z_2 
 \end{bmatrix}
\begin{bmatrix}
  \mathbf{K}_{12}\mathbf{K}_{22}^{-1}\mathbf{W}_2\boldsymbol{\alpha} + \boldsymbol{\epsilon}^{Complete}  \\
  \mathbf{W}_2\boldsymbol{\alpha} \\
 \end{bmatrix}
  + \mathbf{e} \\
    & = 
 \begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix}
 \boldsymbol{\beta} + 
 \begin{bmatrix}
  Z_1 & 0 \\
  0 & Z_2 
 \end{bmatrix}
\begin{bmatrix}
  \hat{\mathbf{W}_1}\boldsymbol{\alpha} + \boldsymbol{\epsilon}^{Complete} \\
  \mathbf{W}_2\boldsymbol{\alpha} \\
 \end{bmatrix}
  + \mathbf{e} \\
\end{align}

The breeding values from that model are:

\begin{equation}
 \label{eq:breeding-values}
\tilde{\mathbf{g}} = 
 \begin{bmatrix}
  \hat{\mathbf{W}_1} \\
  \mathbf{W}_2 
 \end{bmatrix}
 \hat{\boldsymbol{\alpha}}
 + 
 \begin{bmatrix}
  \mathbf{Z}_1 \\
  0
 \end{bmatrix}
 \hat{\boldsymbol{\epsilon}}^{Complete} 
\end{equation}

Our final model looks like this:

\begin{align}
 \label{eq:final-model}
\mathbf{y} &= \mathbf{Xb} +
\sum_{l=1}^{L} \mathbf{Z}_{l}\mathbf{W}_{l} \boldsymbol{\alpha}_{l} + 
\sum_{l=1}^{L} \mathbf{Z}_{l}\mathbf{U}_{l}^{Complete} \boldsymbol{\epsilon}_{l}^{Complete} +
\mathbf{e}.
\end{align}

For the hybrid data, the obtained GCA effects for the inbred lines
represent half their breeding values. The predicted agronomic
performance is:

\begin{equation}
 \label{eq:predicted-performance}
\hat{\mathbf{y}} = \sum_{l=1}^{L} \mathbf{Z}_{l}\tilde{\mathbf{g}}_{l} 
\end{equation}

\subsubsection{Predictive ability and model
validation}\label{predictive-ability-and-model-validation}

For the validation of our predictions, we employed leave-one-out
cross-validation (LOOCV) routines. In the case of the diversity panel
maize inbred lines, LOOCV was performed by using a single genotype as a
hold-out sample, which will subsequently be predicted, while using all
other inbred lines for model training. This process is repeated until
all inbred lines have been used once for testing and \(n - 1\) times for
model training.

For the hybrid data, LOOCV was carried out as follows: Let \(D\) and
\(F\) denote the set of parental inbred lines from the Dent and the
Flint group, respectively. Further, let \(H \cap [D \times F]\) denote
the set of hybrids from crosses between \(D\) and \(F\). The training
set (\(H_{TRN}\)) for the hybrid to be predicted (\(H_{ij}\)) --- where
\(i \in D_{TRN}\) and \(j \in F_{TRN}\) --- was assembled as
\(H_{TRN} = [H \cap (D_{TRN}^{C} \times F_{TRN}^{C})]\) with
\(D_{TRN}^{C} = D \setminus D_{TRN}\) and
\(F_{TRN}^{C} = F \setminus F_{TRN}\).

We judged the performance of each model by looking at its predictive
ability, which is calculated as \(\rho(\mathbf{y}, \mathbf{\hat{y}})\),
where \(\mathbf{\hat{y}}\) is the vector of predicted values from each
LOOCV run. Our confidence in the performance of each model was evalutaed
by computing the coefficient of variation for the predicted values:

\begin{equation}
  \label{eq:CV}
CV = \sqrt{\frac{\sum_{i = 1}^{n} \sigma^{2}_{\hat{y}(i)} / n}{\sum_{i=1}^{n}\hat{y}_{i} / n}},
\end{equation}

where \(n\) denotes the number of genotypes (\emph{i.e.}, inbred lines
or hybrids).

\section{Results}\label{results}

\begin{figure}[htbp]
\centering
\includegraphics{"./tabs_figs/core_fraction_predictive_ability_trend.pdf"}
\caption{Bullshit}
\end{figure}

\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for the set of 211 tropical/subtropical lines from the maize diversity panel.} 
\begin{tabular}{rllllll}
  \toprule
 & 100grainweight & cobweight & Eardiameter & Kernelwidth & Plantheight & Silkingtime \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: no}}\\
snp-mrna & 0.44 (0.0748) & 0.33 (0.1727) & 0.4 (0.0543) & 0.34 (0.0425) & 0.4 (0.0707) & 0.53 (0.0285) \\ 
  snp-none & 0.43 (0.0733) & 0.37 (0.1734) & 0.39 (0.0535) & 0.34 (0.0455) & 0.45 (0.0755) & 0.53 (0.0297) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: yes}}\\
mrna-none & 0.4 (0.0747) & 0.15 (0.1568) & 0.37 (0.0509) & 0.36 (0.0423) & 0.32 (0.0692) & 0.49 (0.0279) \\ 
  snp-none1 & 0.36 (0.0773) & 0.21 (0.1678) & 0.3 (0.0527) & 0.34 (0.0472) & 0.41 (0.078) & 0.45 (0.0302) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}

\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for the set of maize hybrids} 
\begin{tabular}{rllllll}
  \toprule
 & DMY & DMC & FAT & PRO & STA & SUG \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: FALSE}}\\
ped-none & 0.54 (0.0493) & 0.48 (0.0521) & 0.21 (0.0365) & 0.6 (0.0187) & 0.35 (0.0427) & 0.42 (0.117) \\ 
  snp-none & 0.74 (0.0517) & 0.63 (0.0521) & 0.43 (0.0356) & 0.65 (0.0223) & 0.46 (0.0458) & 0.47 (0.1261) \\ 
  ped-snp & 0.71 (0.054) & 0.57 (0.0545) & 0.35 (0.0375) & 0.62 (0.0235) & 0.39 (0.0478) & 0.42 (0.1298) \\ 
  ped-mrna & 0.73 (0.056) & 0.53 (0.0594) & 0.35 (0.0398) & 0.61 (0.0254) & 0.31 (0.0504) & 0.37 (0.1396) \\ 
  snp-mrna & 0.75 (0.0533) & 0.61 (0.0553) & 0.43 (0.0374) & 0.65 (0.0239) & 0.44 (0.048) & 0.44 (0.1342) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: TRUE}}\\
mrna-none & 0.75 (0.0552) & 0.34 (0.0619) & 0.4 (0.0412) & 0.47 (0.0293) & 0.16 (0.0496) & 0.27 (0.1447) \\ 
  ped-none1 & 0.46 (0.0469) & 0.53 (0.05) & 0.23 (0.0368) & 0.46 (0.0234) & 0.38 (0.0391) & 0.34 (0.1316) \\ 
  snp-none1 & 0.67 (0.0534) & 0.47 (0.0552) & 0.43 (0.0384) & 0.48 (0.027) & 0.36 (0.0457) & 0.35 (0.132) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}

\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for different core sets from the tropical/subtropical material of the maize diversity panel.} 
\begin{tabular}{rllllll}
  \toprule
 & 100grainweight & cobweight & Eardiameter & Kernelwidth & Plantheight & Silkingtime \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: FALSE}}\\
snp-mrna & 0.44 (0.0748) & 0.33 (0.1727) & 0.4 (0.0543) & 0.34 (0.0425) & 0.4 (0.0707) & 0.53 (0.0285) \\ 
  snp-none & 0.43 (0.0733) & 0.37 (0.1734) & 0.39 (0.0535) & 0.34 (0.0455) & 0.45 (0.0755) & 0.53 (0.0297) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: TRUE}}\\
mrna-none\_1.0 & 0.4 (0.0747) & 0.15 (0.1568) & 0.37 (0.0509) & 0.36 (0.0423) & 0.32 (0.0692) & 0.49 (0.0279) \\ 
  snp-mrna\_0.3 & 0.36 (0.0711) & 0.2 (0.1566) & 0.28 (0.0503) & 0.37 (0.044) & 0.42 (0.0736) & 0.5 (0.028) \\ 
  snp-mrna\_0.4 & 0.39 (0.0697) & 0.2 (0.1558) & 0.33 (0.0486) & 0.35 (0.0428) & 0.38 (0.0727) & 0.52 (0.028) \\ 
  snp-mrna\_0.5 & 0.41 (0.0671) & 0.16 (0.1549) & 0.31 (0.0486) & 0.4 (0.0414) & 0.44 (0.0748) & 0.49 (0.0278) \\ 
  snp-mrna\_0.6 & 0.4 (0.0739) & 0.13 (0.1563) & 0.27 (0.0497) & 0.37 (0.0417) & 0.42 (0.075) & 0.55 (0.0272) \\ 
  snp-mrna\_0.7 & 0.37 (0.0734) & 0.1 (0.1555) & 0.31 (0.0508) & 0.37 (0.0422) & 0.33 (0.0755) & 0.52 (0.028) \\ 
  snp-mrna\_0.8 & 0.4 (0.0774) & 0.21 (0.1581) & 0.36 (0.0515) & 0.41 (0.0458) & 0.29 (0.0737) & 0.49 (0.0288) \\ 
  snp-mrna\_0.9 & 0.39 (0.0792) & 0.21 (0.1692) & 0.38 (0.0561) & 0.41 (0.0459) & 0.29 (0.075) & 0.49 (0.0311) \\ 
  snp-none\_1.0 & 0.36 (0.0773) & 0.21 (0.1678) & 0.3 (0.0527) & 0.34 (0.0472) & 0.41 (0.078) & 0.45 (0.0302) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}

\section{Discussion}\label{discussion}

\section{References}\label{references}

\hypertarget{refs}{}
\hypertarget{ref-Browning2009}{}
Browning, Brian L, and Sharon R Browning. 2009. ``A unified approach to
genotype imputation and haplotype-phase inference for large data sets of
trios and unrelated individuals.'' \emph{Am. J. Hum. Genet.} 84 (2). The
American Society of Human Genetics: 210--23.
doi:\href{https://doi.org/10.1016/j.ajhg.2009.01.005}{10.1016/j.ajhg.2009.01.005}.

\hypertarget{ref-Fernando2014}{}
Fernando, Rohan L, Jack Cm Dekkers, and Dorian J Garrick. 2014. ``A
class of Bayesian methods to combine large numbers of genotyped and
non-genotyped animals for whole-genome analyses.'' \emph{Genetics,
Selection, Evolution : GSE} 46 (1): 50.
doi:\href{https://doi.org/10.1186/1297-9686-46-50}{10.1186/1297-9686-46-50}.

\hypertarget{ref-Frichot2015}{}
Frichot, Eric, and Olivier François. 2015. ``LEA: An R package for
landscape and ecological association studies.'' \emph{Methods in Ecology
and Evolution} 6 (8): 925--29.
doi:\href{https://doi.org/10.1111/2041-210X.12382}{10.1111/2041-210X.12382}.

\hypertarget{ref-Frichot2014}{}
Frichot, Eric, François Mathieu, Théo Trouillon, Guillaume Bouchard, and
Olivier François. 2014. ``Fast and efficient estimation of individual
ancestry coefficients.'' \emph{Genetics} 196 (4): 973--83.
doi:\href{https://doi.org/10.1534/genetics.113.160572}{10.1534/genetics.113.160572}.

\hypertarget{ref-Frisch2010}{}
Frisch, Matthias, Alexander Thiemann, Junjie Fu, Tobias a. Schrag,
Stefan Scholten, and Albrecht E. Melchinger. 2010. ``Transcriptome-based
distance measures for grouping of germplasm and prediction of hybrid
performance in maize.'' \emph{Theor. Appl. Genet.} 120 (2): 441--50.
doi:\href{https://doi.org/10.1007/s00122-009-1204-1}{10.1007/s00122-009-1204-1}.

\hypertarget{ref-Fu2013}{}
Fu, Junjie, Yanbing Cheng, Jingjing Linghu, Xiaohong Yang, Lin Kang,
Zuxin Zhang, Jie Zhang, et al. 2013. ``RNA sequencing reveals the
complex regulatory network in the maize kernel.'' \emph{Nat. Comm.} 4.
Nature Publishing Group: 2832.
doi:\href{https://doi.org/10.1038/ncomms3832}{10.1038/ncomms3832}.

\hypertarget{ref-Ganal2011}{}
Ganal, Martin W, Gregor Durstewitz, Andreas Polley, Aurélie Bérard,
Edward S Buckler, Alain Charcosset, Joseph D Clarke, et al. 2011. ``A
large maize (Zea mays L.) SNP genotyping array: development and
germplasm genotyping, and genetic mapping to compare with the B73
reference genome.'' \emph{PloS ONE} 6 (12): e28334.
doi:\href{https://doi.org/10.1371/journal.pone.0028334}{10.1371/journal.pone.0028334}.

\hypertarget{ref-Kamvar2015}{}
Kamvar, Zhian N., Jonah C. Brooks, and Niklaus J. Grünwald. 2015.
``Novel R tools for analysis of genome-wide population genetic data with
emphasis on clonality.'' \emph{Frontiers in Genetics} 6: 1--10.
doi:\href{https://doi.org/10.3389/fgene.2015.00208}{10.3389/fgene.2015.00208}.

\hypertarget{ref-Legarra2009}{}
Legarra, a, I Aguilar, and I Misztal. 2009. ``A relationship matrix
including full pedigree and genomic information.'' \emph{Journal of
Dairy Science} 92 (9). Elsevier: 4656--63.
doi:\href{https://doi.org/10.3168/jds.2009-2061}{10.3168/jds.2009-2061}.

\hypertarget{ref-Li2013}{}
Li, Hui, Zhiyu Peng, Xiaohong Yang, Weidong Wang, Junjie Fu, Jianhua
Wang, Yingjia Han, et al. 2013. ``Genome-wide association study dissects
the genetic architecture of oil biosynthesis in maize kernels.''
\emph{Nat. Genet.} 45 (1). Nature Publishing Group: 43--50.
doi:\href{https://doi.org/10.1038/ng.2484}{10.1038/ng.2484}.

\hypertarget{ref-Mrode2014}{}
Mrode, Raphael A. 2014. \emph{Linear Models for the Prediction of Animal
Breeding Values}. 3rd ed. Oxfordshire: CABI.
doi:\href{https://doi.org/10.1017/CBO9781107415324.004}{10.1017/CBO9781107415324.004}.

\hypertarget{ref-Perez2014}{}
Pérez, Paulino, and Gustavo de Los Campos. 2014. ``Genome-wide
regression \& prediction with the BGLR statistical package.''
\emph{Genetics} 198 (October): 483--95.
doi:\href{https://doi.org/10.1534/genetics.114.164442}{10.1534/genetics.114.164442}.

\hypertarget{ref-Ritchie2015a}{}
Ritchie, M. E., B. Phipson, D. Wu, Y. Hu, C. W. Law, W. Shi, and G. K.
Smyth. 2015. ``Limma powers differential expression analyses for
RNA-sequencing and microarray studies.'' \emph{Nucleic Acids Res.} 43
(7): e47.
doi:\href{https://doi.org/10.1093/nar/gkv007}{10.1093/nar/gkv007}.

\hypertarget{ref-Ritchie2007}{}
Ritchie, Matthew E, Jeremy Silver, Alicia Oshlack, Melissa Holmes,
Dileepa Diyagama, Andrew Holloway, and Gordon K Smyth. 2007. ``A
comparison of background correction methods for two-colour
microarrays.'' \emph{Bioinformatics} 23 (20): 2700--2707.
doi:\href{https://doi.org/10.1093/bioinformatics/btm412}{10.1093/bioinformatics/btm412}.

\hypertarget{ref-Smyth2003}{}
Smyth, Gordon K, and Terry Speed. 2003. ``Normalization of cDNA
microarray data.'' \emph{Methods} 31 (4): 265--73.
doi:\href{https://doi.org/10.1016/S1046-2023(03)00155-5}{10.1016/S1046-2023(03)00155-5}.

\hypertarget{ref-Thachuk2009}{}
Thachuk, Chris, José Crossa, Jorge Franco, Susanne Dreisigacker, Marilyn
Warburton, and Guy F Davenport. 2009. ``Core Hunter: an algorithm for
sampling genetic resources based on multiple genetic measures.''
\emph{BMC Bioinformatics} 10 (243): 1--13.
doi:\href{https://doi.org/10.1186/1471-2105-10-243}{10.1186/1471-2105-10-243}.

\hypertarget{ref-VanRaden2008}{}
VanRaden, P M. 2008. ``Efficient methods to compute genomic
predictions.'' \emph{J. Dairy Sci.} 91 (11). Elsevier: 4414--23.
doi:\href{https://doi.org/10.3168/jds.2007-0980}{10.3168/jds.2007-0980}.

\hypertarget{ref-Westhues2017}{}
Westhues, Matthias, Tobias A Schrag, Claas Heuer, Georg Thaller, H.F.
Utz, Wolfgang Schipprack, Alexander Thiemann, et al. 2017. ``Omics-based
Hybrid Prediction in Maize.''

\hypertarget{ref-Yang2014}{}
Yang, Ning, Yanli Lu, Xiaohong Yang, Juan Huang, Yang Zhou, Farhan Ali,
Weiwei Wen, Jie Liu, Jiansheng Li, and Jianbing Yan. 2014. ``Genome Wide
Association Studies Using a New Nonparametric Model Reveal the Genetic
Architecture of 17 Agronomic Traits in an Enlarged Maize Association
Panel.'' \emph{PLoS Genet.} 10 (9).
doi:\href{https://doi.org/10.1371/journal.pgen.1004573}{10.1371/journal.pgen.1004573}.

\end{document}


