# LOOCV-Transformed-Kernel-Prediction
```{r setup-transformed-kernel-prediction, message = FALSE, echo = FALSE}
if (!require("pacman")) install.packages("pacman")
pacman::p_load("tidyverse", "data.table", "dtplyr", "ggthemes", "viridis",
               "stringr", "forcats")

knitr::knit_hooks$set(inline = function(x) {
  prettyNum(x, big.mark = ",", digits = 2)
})
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(fig.width = 8, fig.height = 6,
                      warning = FALSE, echo = FALSE)
```



```{r Load-LOOCV-Data, cache = TRUE}
# Specify which predictions are based on the updated single-step prediction
# algorithms in 'sspredr'.
commit <-  "mwesthues/sspredr@26447a1"
session_files <- list.files("./data/derived/session_info/")
session_files <- session_files[grep(session_files,
                                    pattern = "interactive", invert = TRUE)]
session_filter <- session_files %>%
  map(~readLines(paste0("./data/derived/session_info/", .))) %>%
  grepl(., pattern = commit)
upd_algo_runs <- session_files[session_filter]
upd_algo_runs <- gsub(upd_algo_runs, pattern = ".txt", replacement = "")
rep_log_file <- fread("./data/derived/repeated_snp77_prediction_log.txt") %>%
  filter(Cores == 16,
         Job_ID %in% upd_algo_runs)

# Load all prediction results based on the resampling of 'snp77'.
rep_df <- rep_log_file %>%
  .$Job_ID %>%
  as.list() %>%
  map(~paste0("./data/derived/predictions/", ., ".RDS")) %>%
  map(readRDS) %>%
  bind_rows() %>%
  left_join(y = rep_log_file %>% select(Job_ID, Pred1, Pred2, Pred3, Iter,
                                        Replication),
            by = "Job_ID") %>%
  unite(col = Predictor, Pred1, Pred2, Pred3, sep = "-")


# Load all other predictions, which did not involve 'snp77'.
reg_log <- fread("./data/derived/pred_log.txt") %>%
  select(-Model, -PI, -PriorPiCount) %>%
  filter(Cores == 16,
        # Job_ID %in% upd_algo_runs,
         Transformation == TRUE | Pred1 %in% c("ped42", "ped100"),
         Pred1 != "snp77",
         Pred2 != "snp77",
         Pred3 != "snp77") %>% 
  select(-Transformation)
reg_df <- reg_log %>%
  .$Job_ID %>%
  as.list() %>%
  map(~paste0("./data/derived/predictions/", ., ".RDS")) %>%
  map(readRDS) %>%
  bind_rows() %>%
  left_join(y = reg_log %>% select(Job_ID, Pred1, Pred2, Pred3, Iter),
            by = "Job_ID") %>%
  unite(col = Predictor, Pred1, Pred2, Pred3, sep = "-")

# Combine all data for ease of handling.
pred_df <- rbind(reg_df, rep_df %>% select(-Replication))
pred_df <- pred_df %>%
  mutate(Trait = as.factor(Trait),
         Trait = fct_recode(Trait,
    "DMY" = "GTM",
    "DMC" = "GTS",
    "FAT" = "FETT",
    "PRO" = "RPR",
    "SUG" = "XZ"
  )) %>%
  mutate(Trait = fct_relevel(Trait,
                             "DMY", "DMC", "ADF", "FAT", "PRO", "STA", "SUG"),
         Predictor = as.factor(Predictor)) %>%
  select(-Job_ID, -Iter)
```


```{r Reduce-Hybrid-Data, dependson="Load-Agro-Data"}
red_hybrid <- readRDS("./data/processed/hybrids_685.RDS")
pred_df <- pred_df %>%
  filter(Geno %in% red_hybrid)
```


## Distribution of predicted and observed values
```{r Distribution-y-yhat, cache = TRUE, fig.height = 12, fig.cap = "Frequency polygon of observed and predicted values for seven agronomic traits (columns) and nine sets of predictors (rows)."}
# Specify a unified predictor order.
pred_order <- c("ped42-none-none", "snp42-none-none", "mrna42-none-none",
                "ped100-none-none", "snp100-none-none", "ped100-snp77-none",
                "ped100-mrna42-none", "snp100-mrna42-none",
                "ped100-snp77-mrna42")
legend_labs <- expression(y, hat(y))
pred_df %>%
  filter(Predictor != "snp77-none-none") %>%
  gather(key = Value, value = `Phenotypic Value`, y, yhat) %>%
  mutate(Predictor = fct_relevel(Predictor, pred_order),
         Value = as.factor(Value)) %>%
  ggplot(aes(x = `Phenotypic Value`, color = Value)) +
  geom_freqpoly() +
  scale_color_manual(values = c("#1f78b4", "#e31a1c"), labels = legend_labs) +
  facet_grid(Predictor ~ Trait, scales = "free") +
  theme_bw() +
  theme(legend.position = "top",
        strip.text.y = element_text(size = 7)) +
  guides(color = guide_legend(override.aes = list(size = 2))) +
  ylab("Frequency")
```

## Coefficient of variation for predicted values
We are interested in the variation in the predicted values relative to the 
arithmetic mean of the predicted values, separately for each agronomic trait and
each set of predictors:

$$
\begin{equation}
  \label{eq:CV}
CV = \sqrt{\frac{\sum_{i = 1}^{N} \sigma^{2}_{\hat{y}(i)} / N}{\sum_{i=1}^{N}\hat{y}_{i} / N}},
\end{equation}
$$

where $N$ denotes the number of hybrids.


```{r Coefficient-of-Variation, cache = TRUE}
coefficient_of_variation <- function(x, y) {
  sqrt(mean(x)) / mean(y)
}
cv_df <- pred_df %>%
  split(list(.$Trait, .$Predictor)) %>%
  map(~coefficient_of_variation(x = .$var_yhat, y = .$yhat)) %>%
  stack() %>%
  separate(col = ind, into = c("Trait", "Predictor"), sep = "[.]") %>%
  rename(CV = values) %>%
  tbl_dt() %>%
  mutate(Predictor = as.factor(Predictor),
         Trait = as.factor(Trait))

overview_df <- pred_df %>%
  group_by(Predictor, Trait) %>%
  summarize(r = cor(y, yhat)) %>%
  ungroup() %>%
  as.data.frame() %>%
  left_join(y = cv_df, by = c("Predictor", "Trait")) %>%
  as_tibble() %>%
  mutate(Predictor = as.factor(Predictor),
         Predictor = forcats::fct_reorder(Predictor, CV),
         Trait = fct_relevel(Trait,
                             "DMY", "DMC", "ADF", "FAT", "PRO",
                             "STA", "SUG"),
         Group = ifelse(Predictor %in% c("snp42-none-none",
                                         "ped42-none-none",
                                         "mrna42-none-none"),
                        yes = "Reduced", no = "Full")
  ) %>%
  mutate(Group = as.factor(Group))
```



```{r Coefficient-of-Variation-Plot, cache = TRUE, fig.cap = "Coefficient of variation for seven agronomic traits and nine sets of predictors."}
overview_df %>%
  filter(Predictor != "snp77-none-none") %>%
  mutate(Predictor = fct_relevel(Predictor, pred_order)) %>%
  ggplot(aes(x = CV, y = Predictor)) +
  geom_segment(aes(yend = Predictor), xend = 0, color = "gray50") +
  geom_point(aes(color = Predictor), size = 3) +
  scale_color_viridis(discrete = TRUE, option = "inferno") +
  facet_wrap(~ Trait, scales = "free_x") +
  theme_bw() +
  theme(panel.grid.major.y = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        legend.position = c(0.7, 0.13)
  ) +
  guides(color = guide_legend(ncol = 2,
                              reverse = FALSE)) +
  xlab("Coefficient of Variation") +
  ylab("Predictor Set") +
  xlim(0, NA)
```

## Predictive Abilities
For an introduction to the abbreviations used for the different predictors, and
combinations thereof, please refer to chapter \@ref(predictor-subsetting).

```{r Predictive-Ability-Plots, cache = TRUE, fig.height = 10, fig.cap = "Predictive ability for seven agronomic traits (rows) and nine predictor sets (columns). The predictor sets are split into the categories 'Reduced' (i.e. 685 hybrids) and 'Full' (i.e. 1,521 hybrids). For both categories, predictive abilities are based solely on the set of 685 common hybrids. Predictions including 'snp77' are based on 20 runs where each run was assembled by selecting a set of genotypes at random. The resampling procedure took into account the difference in absolute size between sets of Dent and the Flint parent lines. In these instances, the arithmetic mean of predictive abilities across all 20 runs is displayed."}
overview_df %>%
  as_tibble() %>%
  filter(Predictor != "snp77-none-none") %>%
  mutate(Predictor = fct_relevel(Predictor, pred_order),
         Group = fct_relevel(Group, "Reduced", "Full")
  ) %>%
  ggplot(aes(x = Predictor, y = r)) +
  geom_bar(stat = "identity", aes(fill = Predictor), color = "black") +
  geom_text(aes(label = round(r, digits = 3)), vjust = -0.2) +
  scale_fill_viridis(discrete = TRUE, option = "inferno") +
  facet_grid(Trait ~ Group, space = "free", scales = "free_x") +
  theme_bw() +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.key = element_rect(color = "white")
  ) +
  guides(fill = guide_legend(nrow = 3,
                             byrow = TRUE,
                             keywidth = 0.25,
                             keyheight = 0.25,
                             default.unit = "inch",
                             title.position = "top")
  ) +
  ylim(c(0, 1))
```


### Findings
a)   Reduced
   
   -   For dry matter yield, transcriptomic data clearly perform best.

   -   Genomic data perform decisively better than pedigree data for DMY, ADF 
     and FAT.

   -   Relative to pedigree and genomic data, transcripts perform poorly for 
     DMC and STA.

b)   Full

   -   SNP data perform better for all traits compared to pedigree information.

   -   Predictive abilities for the combination of pedigree, SNP and 
     transcriptomic data "ped100-snp77-mrna42" seem to be the sum of the
     predictive abilities of the individual predictors weighted by their share
     of the genotypes. 
     For example 
     $$r(\text{ped100-snp77-mrna42}) = r(\text{ped100}) * (1 - 0.77) +
     r(\text{snp77}) * (0.77 - 0.42) + r(\text{mrna42}) * 0.42$$

   -   The predictive ability of only two predictors is typically higher than a
     weighted average of the predictive abilities of the constituting
     individual predictors.

   -   Predictive abilities for FAT and the predictor combination
     "ped100-snp77-mrna42" are difficult to explain. **Suggestions???**

c)   General

   -   If a reduced predictor performed better than another predictor in the 
     reduced data set (e.g. 'mrna42' vs. 'snp42' for DMY), it will boost the 
     performance of the other predictor (here 'snp100') in the full set when the 
     two predictors are combined in a single-step approach (here 
     'snp100-mrna-42-none').



#### Predictive abilities for predictor sets with resampling.
```{r Resampled-Predictive-Ability-Plots, cache = TRUE, fig.height = 10, fig.cap = "Predictive ability for two scenarios and seven agronomic traits where the composition of the set of genomic predictors 'snp77' is based on sampling at random for 20 times. Predictive abilities are based solely on the set of 685 common hybrids."}
# Plot the distribution of predictive abilities over 20 rounds of resampled
# 'snp77'.
rep_df %>%
  mutate(Trait = as.factor(Trait),
         Trait = fct_recode(Trait,
    "DMY" = "GTM",
    "DMC" = "GTS",
    "FAT" = "FETT",
    "PRO" = "RPR",
    "SUG" = "XZ"
  )) %>%
  mutate(Trait = fct_relevel(
    Trait, "DMY", "DMC", "ADF", "FAT", "PRO", "STA", "SUG"
    )
  ) %>%
  group_by(Trait, Replication, Predictor) %>%
  summarize(r = cor(y, yhat)) %>%
  ggplot(aes(x = Predictor, y = r, fill = Predictor)) +
  geom_boxplot() +
  facet_wrap(~Trait) +
  theme_bw(base_size = 10) +
  theme(axis.text.x = element_blank(),
        legend.position = "top")
```



