\documentclass[12pt,titlepage]{article}

%% THE USEPACKAGES NECESSARY FOR THIS EXAMPLE
%% NOTE THAT genetics_manu_style MUST BE CALLED AFTER mychicago
\usepackage{graphicx}
\usepackage{endfloat}
\usepackage{amsfonts}
\usepackage{mychicago}
\usepackage{subfigure}
\usepackage{genetics_manu_style}
\usepackage{authblk} % improved citations
\usepackage{setspace} %double spacing
\usepackage{lineno} % line numbers
\usepackage{booktabs} % book-quality tables
\usepackage{multicol} % list corresponding authors in two separate columns

%% Probably unnecessary packages.
\usepackage[T1]{fontenc}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript




%% THE MANUSCRIPT TITLE
\title{Omics-based Single Step Trait Prediction}


% Allow for multiple authors to share the same institution.
\newcommand*\samethanks[1][\value{footnote}]{\footnotemark[#1]}

\author{
  Matthias Westhues\thanks{Institute of Plant Breeding, Seed Science and Population Genetics, University of Hohenheim, 70599 Stuttgart, Germany},
  Claas Heuer\thanks{Institute of Animal Breeding and Husbandry, Christian-Albrechts-University Kiel, 24098 Kiel, Germany},
  Georg Thaller\samethanks[2],
  Rohan Fernando\thanks{Department of Animal Science, Iowa State University, 50011 Ames, Iowa, U.S.A.},
  Albrecht E. Melchinger\samethanks[1]
}


\renewcommand{\CorrespondingAddress}{%
    University of Hohenheim \\
    Institute of Plant Breeding, Seed Science and Population Genetics \\ 
    Fruwirthstr. 21 \\ 
    Stuttgart 70599, GERMANY \\
    Tel.: (+49) 0711 459-22334 \\
    Fax: (+49) 0711 459-22343 \\ 
    \texttt{melchinger@uni-hohenheim.de} \vfill}

\renewcommand{\RunningHead}{Single Step Prediction}
\renewcommand{\CorrespondingAuthor}{Prof. Dr. A. E. Melchinger}
\renewcommand{\KeyWords}{single step, omics, prediction, maize}




%% SOME COMMANDS FOR THE CONTENT OF THIS FILE.  NOT NECESSARY FOR
%% GENETIC_MANU_STYLE
\newcommand{\bp}{\mathbf{p}}
\newcommand{\LLL}{\mathcal{L}}




%% BEGIN DOC
\begin{document}


\maketitle
\doublespacing
\linenumbers



\begin{abstract}

  $\dots$

\end{abstract}



% Please add here a significance statement to explain the relevance of your work
\section{Article Summary}
$\dots$



% Introduction
\section{Introduction} 

$\dots$

Our objectives were threefold: (1) Examine whether the single step prediction
framework can be transfered to hybrids, (2) explore the utility of single step 
predition with a quantitative predictor and (3) evaluate the impact of the
particular genetic space covered by the incomplete predictor on predictive
ability.




\section{Materials and Methods}
\Genetics2level{Maize hybrids}
The maize hybrid data set comprises 1,521 hybrids produced in 16 factorial
mating designs between 142 Dent and 103 Flint parent lines \cite{Westhues2017}. 
Best linear unbiased estimates (BLUEs) were computed across all factorials and 
three or more agro-ecologically diverse locations across Germany for six 
agronomically important traits.
All 245 parent lines have pedigree records back to the generation of their 
grandparents or deeper \cite{Westhues2017} and were genotyped using the Illumina 
BeadChip MaizeSNP50 \cite{Ganal2011}.
After filtering for minor allele frequency ($\geq 5$\%), heterozygosity rate
($\leq 5$\%) and call frequency ($\geq 95$\%), missing genotypes were imputed 
using the Beagle software \cite{Browning2009}.
The final number of polymorphic marker loci was 7,013 for parental Dent and
6,212 for parental Flint lines, respectively.
Gene expression data for seedlings seven days after sowing were obtained for a 
subset of 60 parental Dent and 43 parental Flint lines through two-color
hybridizations using a custom 2K microarray (GPL22267) \cite{Westhues2017}.
After applying established normalization procedures 
\cite{Smyth2003,Ritchie2007}, gene expression BLUEs were computed for 1,323 
transcripts in reference to established protocols 
\cite{Smyth2003,Ritchie2007,Frisch2010} using the R-package \textit{limma}
\cite{Ritchie2015a}.



\Genetics2level{Maize diversity panel}
The maize inbred line data set originally comprised 513 maize lines representing 
the global maize diversity and was reduced to the set of tropical and
subtropical lines ($n = 211$) given that it is the largest of the four
pre-classified subgroups \cite{Yang2014}.
All inbred lines were evaluated in five Chinese environments ranging from
$18$ to $30^{\circ}$N and from $102$ to $110^{\circ}$E \cite{Yang2014}.
Best linear unbiased predictors (BLUPs) were calculated for 17 traits of which 
six were used for this study.
MaizeSNP50 BeadChip-data were available for the entire set of lines from the
diversity panel.
Additionally, RNA sequencing was performed for 368 lines on young seedlings 15 
days after sowing.
By exploting identify by descent (IBD) between 49,728 SNPs on the BeadChip 
overlapping with the RNA-seq data, 556,809 high quality SNPs could be inferred 
for these lines \cite{Fu2013,Li2013}.
For the remaining 145 maize lines without RNA-seq data, high density markers
were inferred via projection of IBD regions onto the BeadChip, using this core 
set of markers \cite{Yang2014}.
The same SNP quality checks as for the parental inbred lines of the hybrids
were applied to the 211 inbred lines from the tropical/subtropical subset,
yielding 37,760 SNPs.
Gene expression data were normalized using a normal quantile transformation to
satisfy modelling assumptions \cite{Fu2013}.
A subset of 28,850 annotated genes was available for 149 out of the 211
genotypes and was kept for further analyses \cite{Li2013}.



\Genetics2level{Population structure}
Principal component analyses (PCA) were run on scaled and centered predictor 
matrices using the \emph{pca} function from the R-package \textit{LEA} 
\cite{Frichot2015}.
Estimates of ancestry coefficients were estimated based on sparse nonnegative
matrix factorization algorithms \cite{Frichot2014} using the \emph{snmf} 
function in \textit{LEA}.
Algorithms were run for $K=2, \dots 5$ putative ancestral gene pools with 25
repetitions for each value of $K$. 
For each $K$, only the repetition with the lowest cross-entropy was kept for 
further analyses.



\Genetics2level{Core sampling}
In order to evaluate the influence of the genetic constitution of the set of
genotypes that has only data on one out of two predictors, we generated core 
samples using the R-package \textit{corehunter} \cite{Thachuk2009}.
These core samples were based on the 149 maize inbred lines from the diversity
panel which had gene expression as well as genomic information.
The size of the core set was varied in increments of ten percentage points and 
ranged from 10\% to 90\% of all 149 inbred lines.
For assembling the various core sets, we first computed the Modifed Rogers (MR)
distance between each pair of genotypes and then selected a predefined subset
of genotypes that display maximum average distance between each other.
MR was computed using the \emph{rogers.dist} function from the R-pacakge
\textit{poppr} \cite{Kamvar2015}.


\Genetics2level{Prediction}
\subsection{Kernels}
BLUP models are computationally attractive when the number of features $p$
exceeds the number of genotypes $n$ and are equivalent to a selection index
when fixed effects have been accomodated by the dependent variable 
\cite{Mrode2014}.
Depending on the data set, up to three predictors were available for agronomic
trait predictions, namely pedigree data (P), genomic data (G) and
transcriptompic data (T).
The corresponding feature matrix of the $l$-th group of inbred lines and the 
$o$-th predictor ($\mathbf{W}_{lo}$) has dimensions $n \times p$ where $n$
pertains to the number of genotypes in the $l$-th group and $p$ pertains to 
the number of features.
For the diversity panel maize lines, $n$ corresponds to the number of genotypes
whereas, in the case of the hybrid data, $n$ corresponds to the number of
parent lines in the corresponding heterotic group.
In generating kernels for each predictor, all features in $\mathbf{W}_{lo}$ were
centered and standardized to unit variance.
Then, each kernel can be defined as

\begin{equation} \label{eq:GenomicRelationship}
  \mathbf{K}_{lo} = \frac{1}{W_{lo}} \mathbf{W}_{lo} \mathbf{W}_{lo}^{\top},
\end{equation}

where $W_{lo}$ denotes the number of features \cite{VanRaden2008}.
In the case of pedigree data (P), coancestry coefficients were used directly
for $\mathbf{K}_{lo}$.





\subsection{Universal breeding value model}
The universal model for breeding values was:

\begin{equation} \label{eq:KBLUPModel}
  \mathbf{y} = \mu + 
  \sum_{l=1}^{L} \mathbf{Z}_{lo} \mathbf{g}_{lo} +
  \mathbf{\epsilon},
\end{equation}


where $\mathbf{y}$ is the vector of observed inbred line or hybrid performance,
respectively, $\mu$ is the fixed model intercept, $\mathbf{Z}_{lo}$ is the 
corresponding design matrix associating the random genotype effects 
(\textit{i.e.}, breeding values) of the lines in the $l$-th inbred line group 
and the $o$-th predictor type with $\mathbf{y}$.
Note that, in the case of hybrid data, $\mathbf{Z}_{lo}$ associates the random
general combining ability effects in the parental inbred lines (either Dent or
Flint, as expressed by $l$) with the vector of observed hybrid performance
$\mathbf{y}$.
The random effects ($\mathbf{g}_{lo}$) have expectation zero and variance equal 
to $\mathbf{K}_{lo} \sigma^{2}_{{g}^{lo}}$ and $\mathbf{I} \sigma^2_{\epsilon}$ 
for the residual error.
All prediction models were implemented using the R-package \textit{BGLR}
\cite{Perez2014}. 





\subsection{Single step prediction}
The breeding values in a prediction model are in general given by:
\begin{equation} \label{eq:mrnaebv}
	\mathbf{\tilde{g}} = \mathbf{W}\boldsymbol{\hat{\alpha}},
\end{equation}

where $\boldsymbol{\hat{\alpha}}$ is the solution to a ridge regression model,
which is equivalent to the BLUP model.

Consider now the situation in which one predictor is complete in the sense that
it contains information on the whole population whereas a second predictor is
incomplete and covers only a subset of the population.
We can take a similar route as in \citeNP{Fernando2014} and impute covariates 
of the incomplete predictor by using covariates of the complete predictor.
Let the subscript $1$ denote individuals covered only by the complete
predictor whereas individuals covered by both, the complete and the
incomplete predictor, are indicated by the subscript $2$.
Further, the covariates in $\mathbf{W}$ are centered.
The vector $\mathbf{g}_1$ can be written as the sum of the conditional 
expectation given $\mathbf{g}_2$ and a residual error term:
\begin{align} \label{eq:mrna1}
	\mathbf{g_1} &= E(\mathbf{g}_1|\mathbf{g}_2) + \boldsymbol{\epsilon} \\
	&= \mathbf{K_{12}}\mathbf{K_{22}}^{-1}\mathbf{W_2}\boldsymbol{\hat{\alpha}} + (\mathbf{g_1} - \mathbf{K_{12}}\mathbf{K_{22}}^{-1}\mathbf{W_2}\boldsymbol{\hat{\alpha}}) \\
	&= \mathbf{\hat{g}}_1 + \boldsymbol{\epsilon}
\end{align}

The covariance matrix of $\boldsymbol{\epsilon}$ is $\mathbf{K}_{11} - \mathbf{K}_{12}\mathbf{K}_{22}^{-1}\mathbf{K}_{21} = (\mathbf{K}^{11})^{-1}$
\cite{Legarra2009}.

The structure of the residual imputation/prediction error is known and can 
therefore be modelled.
The covariates not covered by the incomplete predictor can be predicted using 
the expectation of a multivariate normal random vector given correlated 
observations, which are related to the Best Linear Predictor.

The general model for the single step procedure can then be written as 

\begin{equation} \label{eq:single-step-model}
\mathbf{y} = \mathbf{Xb} + \mathbf{W} \boldsymbol{\alpha} + \mathbf{U}^{Complete} \boldsymbol{\epsilon}^{Complete} + \mathbf{e},
\end{equation}

with

\begin{equation} \label{eq:single-step-submatrices}
\mathbf{X} = 
\begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix},
 \mathbf{W} = 
\begin{bmatrix}
  Z_1\hat{\mathbf{W}_1} \\
  \mathbf{W}_2 
 \end{bmatrix},
 \mathbf{U}^{Complete} = 
\begin{bmatrix}
  Z_1 \\
  0 
 \end{bmatrix}
\end{equation}

Following the notation in \citeNP{Fernando2014} the phenotypes can be untangled 
like this:

\begin{align} \label{eq:entangled-augmented-single-step-model}
\begin{bmatrix}
  y_1 \\
  y_2 
 \end{bmatrix}
& =
 \begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix}
 \boldsymbol{\beta} + 
 \begin{bmatrix}
  Z_1 & 0 \\
  0 & Z_2 
 \end{bmatrix}
\begin{bmatrix}
  g_1 \\
  g_2 
 \end{bmatrix}
  + \mathbf{e} \\
    & = 
 \begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix}
 \boldsymbol{\beta} + 
 \begin{bmatrix}
  Z_1 & 0 \\
  0 & Z_2 
 \end{bmatrix}
\begin{bmatrix}
  \mathbf{K}_{12}\mathbf{K}_{22}^{-1}\mathbf{W}_2\boldsymbol{\alpha} + \boldsymbol{\epsilon}^{Complete}  \\
  \mathbf{W}_2\boldsymbol{\alpha} \\
 \end{bmatrix}
  + \mathbf{e} \\
    & = 
 \begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix}
 \boldsymbol{\beta} + 
 \begin{bmatrix}
  Z_1 & 0 \\
  0 & Z_2 
 \end{bmatrix}
\begin{bmatrix}
  \hat{\mathbf{W}_1}\boldsymbol{\alpha} + \boldsymbol{\epsilon}^{Complete} \\
  \mathbf{W}_2\boldsymbol{\alpha} \\
 \end{bmatrix}
  + \mathbf{e} \\
\end{align}

The breeding values from that model are:

\begin{equation} \label{eq:breeding-values}
\tilde{\mathbf{g}} = 
 \begin{bmatrix}
  \hat{\mathbf{W}_1} \\
  \mathbf{W}_2 
 \end{bmatrix}
 \hat{\boldsymbol{\alpha}}
 + 
 \begin{bmatrix}
  \mathbf{Z}_1 \\
  0
 \end{bmatrix}
 \hat{\boldsymbol{\epsilon}}^{Complete} 
\end{equation}



Our final model looks like this:

\begin{align} \label{eq:final-model}
\mathbf{y} &= \mathbf{Xb} +
\sum_{l=1}^{L} \mathbf{Z}_{l}\mathbf{W}_{l} \boldsymbol{\alpha}_{l} + 
\sum_{l=1}^{L} \mathbf{Z}_{l}\mathbf{U}_{l}^{Complete} \boldsymbol{\epsilon}_{l}^{Complete} +
\mathbf{e}.
\end{align}

For the hybrid data, the obtained GCA effects for the inbred lines represent 
half their breeding values.
The predicted agronomic performance is:

\begin{equation} \label{eq:predicted-performance}
\hat{\mathbf{y}} = \sum_{l=1}^{L} \mathbf{Z}_{l}\tilde{\mathbf{g}}_{l} 
\end{equation}




\subsection{Predictive ability and model validation}
For the validation of our predictions, we employed leave-one-out
cross-validation (LOOCV) routines.
In the case of the diversity panel maize inbred lines, LOOCV was performed by 
using a single genotype as a hold-out sample, which will subsequently be 
predicted, while using all other inbred lines for model training.
This process is repeated until all inbred lines have been used once for testing
and $n - 1$ times for model training.

For the hybrid data, LOOCV was carried out as follows:
Let $D$ and $F$ denote the set of parental inbred lines from the Dent and the 
Flint group, respectively.
Further, let $H \cap [D \times F]$ denote the set of hybrids from crosses 
between $D$ and $F$.
The training set ($H_{TRN}$) for the hybrid to be predicted ($H_{ij}$) --- 
where $i \in D_{TRN}$ and $j \in F_{TRN}$ --- was assembled as 
$H_{TRN} = [H \cap (D_{TRN}^{C} \times F_{TRN}^{C})]$ with
$D_{TRN}^{C} = D \setminus D_{TRN}$ and $F_{TRN}^{C} = F \setminus F_{TRN}$.

We judged the performance of each model by looking at its predictive ability,
which is calculated as $\rho(\mathbf{y}, \mathbf{\hat{y}})$, where 
$\mathbf{\hat{y}}$ is the vector of predicted values from each LOOCV run.
Our confidence in the performance of each model was evalutaed by computing the
coefficient of variation for the predicted values:

\begin{equation} \label{eq:CV}
CV = \sqrt{\frac{\sum_{i = 1}^{n} \sigma^{2}_{\hat{y}(i)} / n}{\sum_{i=1}^{n}\hat{y}_{i} / n}},
\end{equation}

where $n$ denotes the number of genotypes (*i.e.*, inbred lines or hybrids).












\section*{Results}

\begin{figure}[H]
\centering
\includegraphics{./tables_figures/core_fraction_predictive_ability_trend.pdf}
\caption{
Influence of core set composition on predictive abilities.
Predictive abilities are based on single step predition with genomic data as the 
complete predictor and transcriptomic data as the incomplete predictor.
Results are shown for six agronomic traits and nine core sets of the maize 
diversity panel.
}
\label{fig:CoreSetInfluence}
\end{figure}


\begin{figure}[H]
\centering
\includegraphics{./tables_figures/uhoh_dent_flint_pca.pdf}
\caption{
  Principal component (PC) analysis of the two heterotic groups of the maize 
  hybrids.
  All inbred lines, which are covered by genomic and transcriptomic features
  are depicted as orange triangles whereas genotypes for which only genomic
  information are available are depicted as blue dots.
}
\label{fig:UHOH-PCA}
\end{figure}


\begin{figure}[H]
\centering
\includegraphics{./tables_figures/maizego_admixture_pca.pdf}
\caption{
  (\textbf{A}) Ancestry coefficients for 211 tropical/subtropical inbred lines
  from the maize diversity panel considering $K=2 \dots 5$ ancestral
  populations.
  (\textbf{B}) Principal component analysis for nine different core sets of the
  set of tropical/subtropical maize inbred lines.
  For each size of the core set, its members are plotted as orange dots whereas 
  all genotypes that are not elements of the cores sets (\textit{i.e.},
  'Complement') are depicted as blue dots.
}
\label{fig:PopStructure}
\end{figure}


% latex table generated in R 3.3.3 by xtable 1.8-2 package
% Wed Mar 15 17:33:34 2017
% Script: ./analysis/maizego_plot_predictive_abilities.R
\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for the set of 211 tropical/subtropical lines from the maize diversity panel.} 
\begin{tabular}{rllllll}
  \toprule
 & 100grainweight & cobweight & Eardiameter & Kernelwidth & Plantheight & Silkingtime \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: FALSE}}\\
G & 0.43 (0.07) & 0.37 (0.17) & 0.39 (0.05) & 0.34 (0.05) & 0.45 (0.08) & 0.53 (0.03) \\ 
  GT & 0.44 (0.07) & 0.33 (0.17) & 0.4 (0.05) & 0.34 (0.04) & 0.4 (0.07) & 0.53 (0.03) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: TRUE}}\\
G & 0.36 (0.08) & 0.21 (0.17) & 0.3 (0.05) & 0.34 (0.05) & 0.41 (0.08) & 0.45 (0.03) \\ 
  T & 0.4 (0.07) & 0.15 (0.16) & 0.37 (0.05) & 0.36 (0.04) & 0.32 (0.07) & 0.49 (0.03) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}




% latex table generated in R 3.3.3 by xtable 1.8-2 package
% Wed Mar 15 17:31:28 2017
% Script: ./analysis/uhoh_plot_predictive_abilities.R
\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for the set of maize hybrids} 
\begin{tabular}{rllllll}
  \toprule
 & DMY & DMC & FAT & PRO & STA & SUG \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: FALSE}}\\
T & 0.75 (0.0552) & 0.34 (0.0619) & 0.4 (0.0412) & 0.47 (0.0293) & 0.16 (0.0496) & 0.27 (0.1447) \\ 
  P & 0.54 (0.0493) & 0.48 (0.0521) & 0.21 (0.0365) & 0.6 (0.0187) & 0.35 (0.0427) & 0.42 (0.117) \\ 
  G & 0.74 (0.0517) & 0.63 (0.0521) & 0.43 (0.0356) & 0.65 (0.0223) & 0.46 (0.0458) & 0.47 (0.1261) \\ 
  PG & 0.71 (0.054) & 0.57 (0.0545) & 0.35 (0.0375) & 0.62 (0.0235) & 0.39 (0.0478) & 0.42 (0.1298) \\ 
  PT & 0.73 (0.056) & 0.53 (0.0594) & 0.35 (0.0398) & 0.61 (0.0254) & 0.31 (0.0504) & 0.37 (0.1396) \\ 
  GT & 0.75 (0.0533) & 0.61 (0.0553) & 0.43 (0.0374) & 0.65 (0.0239) & 0.44 (0.048) & 0.44 (0.1342) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: TRUE}}\\
P & 0.46 (0.0469) & 0.53 (0.05) & 0.23 (0.0368) & 0.46 (0.0234) & 0.38 (0.0391) & 0.34 (0.1316) \\ 
  G & 0.67 (0.0534) & 0.47 (0.0552) & 0.43 (0.0384) & 0.48 (0.027) & 0.36 (0.0457) & 0.35 (0.132) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}





% latex table generated in R 3.3.3 by xtable 1.8-2 package
% Wed Mar 15 17:36:40 2017
% Script: ./analysis/maizego_core_fraction_plot_predictive_abilities.R
\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for different core sets from the tropical/subtropical material of the maize diversity panel.} 
\begin{tabular}{rllllll}
  \toprule
 & 100grainweight & cobweight & Eardiameter & Kernelwidth & Plantheight & Silkingtime \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: FALSE}}\\
G & 0.43 (0.0733) & 0.37 (0.1734) & 0.39 (0.0535) & 0.34 (0.0455) & 0.45 (0.0755) & 0.53 (0.0297) \\ 
  GT & 0.44 (0.0748) & 0.33 (0.1727) & 0.4 (0.0543) & 0.34 (0.0425) & 0.4 (0.0707) & 0.53 (0.0285) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: TRUE}}\\
G\_1.0 & 0.36 (0.0773) & 0.21 (0.1678) & 0.3 (0.0527) & 0.34 (0.0472) & 0.41 (0.078) & 0.45 (0.0302) \\ 
  GT\_0.3 & 0.36 (0.0711) & 0.2 (0.1566) & 0.28 (0.0503) & 0.37 (0.044) & 0.42 (0.0736) & 0.5 (0.028) \\ 
  GT\_0.4 & 0.39 (0.0697) & 0.2 (0.1558) & 0.33 (0.0486) & 0.35 (0.0428) & 0.38 (0.0727) & 0.52 (0.028) \\ 
  GT\_0.5 & 0.41 (0.0671) & 0.16 (0.1549) & 0.31 (0.0486) & 0.4 (0.0414) & 0.44 (0.0748) & 0.49 (0.0278) \\ 
  GT\_0.6 & 0.4 (0.0739) & 0.13 (0.1563) & 0.27 (0.0497) & 0.37 (0.0417) & 0.42 (0.075) & 0.55 (0.0272) \\ 
  GT\_0.7 & 0.37 (0.0734) & 0.1 (0.1555) & 0.31 (0.0508) & 0.37 (0.0422) & 0.33 (0.0755) & 0.52 (0.028) \\ 
  GT\_0.8 & 0.4 (0.0774) & 0.21 (0.1581) & 0.36 (0.0515) & 0.41 (0.0458) & 0.29 (0.0737) & 0.49 (0.0288) \\ 
  GT\_0.9 & 0.39 (0.0792) & 0.21 (0.1692) & 0.38 (0.0561) & 0.41 (0.0459) & 0.29 (0.075) & 0.49 (0.0311) \\ 
  T\_1.0 & 0.4 (0.0747) & 0.15 (0.1568) & 0.37 (0.0509) & 0.36 (0.0423) & 0.32 (0.0692) & 0.49 (0.0279) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}







\Genetics2level{Some results section}

$\dots$

\Genetics2level{Another results section}

$\dots$




\section*{Discussion}
\Genetics2level{Some discussion section}

$\dots$

\subsection{More detailed discussion}

$\dots$

\Genetics2level{Another discussion section}

$\dots$



\section*{Conclusions}

$\dots$




\section{Acknowledgments} 
This project was funded by the German Federal Ministry of Education and 
Research (BMBF) within the projects OPTIMAL (FKZ: 0315958B,0315958F),
SYNBREED (FKZ: 0315528D) and by the German Research Foundation 
(DFG, Grants No. ME 2260/5-1 and SCHO 764/6-1).
The authors acknowledge support by the state of Baden-W{\"u}rttemberg through 
bwHPC.
Financial support for M.W. was provided by the Fiat Panis foundation, Ulm,
Germany.



\nolinenumbers
% Bibliography
\bibliography{library}
\bibliographystyle{mychicago}
\end{document}
