\documentclass[12pt,titlepage]{article}

%% THE USEPACKAGES NECESSARY FOR THIS EXAMPLE
%% NOTE THAT genetics_manu_style MUST BE CALLED AFTER mychicago
\usepackage{graphicx}
\usepackage{endfloat}
\usepackage{amsfonts}
\usepackage{mychicago}
\usepackage{subfigure}
\usepackage{genetics_manu_style}
\usepackage{authblk} % improved citations
\usepackage{setspace} %double spacing
\usepackage{lineno} % line numbers
\usepackage{booktabs} % book-quality tables
\usepackage{multicol} % list corresponding authors in two separate columns

%% Probably unnecessary packages.
\usepackage[T1]{fontenc}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript




%% THE MANUSCRIPT TITLE
\title{Omics-based Single Step Trait Prediction}


% Allow for multiple authors to share the same institution.
\newcommand*\samethanks[1][\value{footnote}]{\footnotemark[#1]}

\author{
  Matthias Westhues\thanks{Institute of Plant Breeding, Seed Science and Population Genetics, University of Hohenheim, 70599 Stuttgart, Germany},
  Claas Heuer\thanks{Institute of Animal Breeding and Husbandry, Christian-Albrechts-University Kiel, 24098 Kiel, Germany},
  Georg Thaller\samethanks[2],
  Rohan Fernando\thanks{Department of Animal Science, Iowa State University, 50011 Ames, Iowa, U.S.A.},
  Albrecht E. Melchinger\samethanks[1]
}


\renewcommand{\CorrespondingAddress}{%
    University of Hohenheim \\
    Institute of Plant Breeding, Seed Science and Population Genetics \\ 
    Fruwirthstr. 21 \\ 
    Stuttgart 70599, GERMANY \\
    Tel.: (+49) 0711 459-22334 \\
    Fax: (+49) 0711 459-22343 \\ 
    \texttt{melchinger@uni-hohenheim.de} \vfill}

\renewcommand{\RunningHead}{Single Step Prediction}
\renewcommand{\CorrespondingAuthor}{Prof. Dr. A. E. Melchinger}
\renewcommand{\KeyWords}{single step, omics, prediction, maize}




%% SOME COMMANDS FOR THE CONTENT OF THIS FILE.  NOT NECESSARY FOR
%% GENETIC_MANU_STYLE
\newcommand{\bp}{\mathbf{p}}
\newcommand{\LLL}{\mathcal{L}}




%% BEGIN DOC
\begin{document}


\maketitle
\doublespacing
\linenumbers



\begin{abstract}

  $\dots$

\end{abstract}



% Please add here a significance statement to explain the relevance of your work
\section{Article Summary}
$\dots$



% Introduction
\section{Introduction} 
% Necessity for predictions.
Genomic prediction, pioneered by \citeNP{Meuwissen2001}, has revolutionized
multiple areas of genetic analyses such as animal and plant breeding
\cite{DeLosCampos2013} as well as disease diagnostics in humans 
\cite{DeLosCampos2010,Vazquez2016}.
The primary application of genomic prediction in breeding programs is for the 
accelerated selection of promising candidates without phenotypes.

% -   Describe the two step procedure and its shortcomings.
% Incomplete coverage of individuals with genomic data.
In animal breeding, only a small fraction of all animals that are recorded in
the pedigree also have genotypes.
To utilize all available phenotpyic information for model training methods were 
needed that would allow for a combined use of genomic and pedigree information 
\cite{Hayes2009a,VanRaden2009}.

% Two-step prediction.
In the early years of genomic prediction, a two-step procedure was the most
intuitive choice for establishing a new selection scheme.
The first step constists of conventionally estimating breeding values based on
pedigree relationships and therefore the phenotypes of relatives.
Individuals with accurate breeding values out of the first step (\textit{i.e.}
high number of offspring), enter the second step in which conventional breeding
values are regressed onto SNP genotypes. 
Those markers can then be used to predict genomic breeding values for solely
genotyped and potentially very young individuals.

% Shortcomings of the two-step procedure.
\begin{itemize}
  \item Bias introduced by genomic information giving an advantage to young,
    genotyped individuals over old animals with only pedigree records
    \cite{Aguilar2009}.
  \item Pedigree depth varies among individuals; some animals have more
    ancestors and therefore more precise estimates \cite{Aguilar2009}.
\end{itemize}

% Problems with SS-BLUP
\begin{itemize}
  \item Weight $w$ for $\mathbf{A}$ and $\mathbf{G}$.
  \item The optimal weight differs between traits \cite{Ashraf2016}.
  \item compatibility between $\mathbf{A}$ and $\mathbf{G}$
    \cite{Christensen2012}
\end{itemize}

% Workings and advantages of the single step procedure.
Independently from each other, \citeNP{Legarra2009} and
\citeNP{Christensen2010} developed a BLUP method that blends the numerator
relationship matrix $\mathbf{A}$ with the genomic relationship matrix
$\mathbf{G}$ in a mutual matrix $\mathbf{H}$.

% Physiological epistasis (added benefits of downstream 'omics' predictors)
Compared to pedigree-based prediction, genomic prediction captures Mendelian
sampling, which is the difference between the expected and the realized
relationship between individuals.
Despite its great success and benefits, genomic prediction does not seem to
capture all physiological epistasis \cite{Sackton2016,Reif2015}, which rather 
manifests itself in the form of main and interaction effects of and between 
downstream 'omics' strata, such as the transcriptome, the proteome and the 
metabolome.
This has lead to the exploration of 

Besides genomic data, other predictors have proven their value in the
prediction of complex traits in plant populations \cite{Guo2016,Westhues2017}
and in humans \cite{Vazquez2016}.

% Special situation in hybrid breeding.
Hybrid breeding --- with noteworthy commercial applications in pig, chicken and 
maize breeding --- is a particularly challenging application field for
prediction tools.
Here, $2n$ parent individuals from two genetically distinct heterotic groups are
crossed to each other, yielding $n^{2}$ potential hybrid progeny that would
require intensive field testing.
In medium-sized plant breeding programs, the advent of the doubled haploid (DH) 
technology \cite{Wedzony2009} now allows for an annual production of 1,000 
parent lines in each heterotic group, amounting to $10^{6}$ putative hybrid
progeny.


% -   Mention scarcity of data on single step prediction in plant populations.
Single step prediction has been a hot topic in animal breeding for a few years,
but so far, to the best of our knowledge, has only been considered in plants
only in one study on wheat lines from CIMMYT \cite{Ashraf2016}. 
So far, its applications in hybrid populations are limited to pig breeding
\cite{Xiang2015,Xiang2016,Tusell2016}.
% Are there applications in chicken breeding? (Hints in Christensen et al.
% 2014, GSE).
Our objectives were threefold: (1) Examine whether the single step prediction
framework can be transfered to hybrids, (2) explore the utility of single step 
predition with a quantitative predictor and (3) evaluate the impact of the
particular genetic space covered by the incomplete predictor on predictive
ability.

% -   How extensive are pedigree records in commercial breeding companies?









\section{Materials and Methods}
\Genetics2level{Maize hybrids}
The maize hybrid data set comprises 1,521 hybrids produced in 16 factorial
mating designs between 142 Dent and 103 Flint parent lines \cite{Westhues2017}. 
Best linear unbiased estimates (BLUEs) were computed across all factorials and 
three or more agro-ecologically diverse locations across Germany for six 
agronomically important traits.
All 245 parent lines have pedigree records back to the generation of their 
grandparents or deeper \cite{Westhues2017} and were genotyped using the Illumina 
BeadChip MaizeSNP50 \cite{Ganal2011}.
After filtering for minor allele frequency ($\geq 5$\%), heterozygosity rate
($\leq 5$\%) and call frequency ($\geq 95$\%), missing genotypes were imputed 
using the Beagle software \cite{Browning2009}.
The final number of polymorphic marker loci was 7,013 for parental Dent and
6,212 for parental Flint lines, respectively.
Gene expression data for seedlings seven days after sowing were obtained for a 
subset of 60 parental Dent and 43 parental Flint lines through two-color
hybridizations using a custom 2K microarray (GPL22267) \cite{Westhues2017}.
After applying established normalization procedures 
\cite{Smyth2003,Ritchie2007}, gene expression BLUEs were computed for 1,323 
transcripts in reference to established protocols 
\cite{Smyth2003,Ritchie2007,Frisch2010} using the R-package \textit{limma}
\cite{Ritchie2015a}.



\Genetics2level{Maize diversity panel}
The maize inbred line data set originally comprised 513 maize lines representing 
the global maize diversity and was reduced to the set of tropical and
subtropical lines ($n = 211$) given that it is the largest of the four
pre-classified subgroups \cite{Yang2014}.
All inbred lines were evaluated in five Chinese environments ranging from
$18$ to $30^{\circ}$N and from $102$ to $110^{\circ}$E \cite{Yang2014}.
Best linear unbiased predictors (BLUPs) were calculated for 17 traits of which 
six were used for this study.
MaizeSNP50 BeadChip-data were available for the entire set of lines from the
diversity panel.
Additionally, RNA sequencing was performed for 368 lines on young seedlings 15 
days after sowing.
By exploting identify by descent (IBD) between 49,728 SNPs on the BeadChip 
overlapping with the RNA-seq data, 556,809 high quality SNPs could be inferred 
for these lines \cite{Fu2013,Li2013}.
For the remaining 145 maize lines without RNA-seq data, high density markers
were inferred via projection of IBD regions onto the BeadChip, using this core 
set of markers \cite{Yang2014}.
The same SNP quality checks as for the parental inbred lines of the hybrids
were applied to the 211 inbred lines from the tropical/subtropical subset,
yielding 37,760 SNPs.
Gene expression data were normalized using a normal quantile transformation to
satisfy modelling assumptions \cite{Fu2013}.
A subset of 28,850 annotated genes was available for 149 out of the 211
genotypes and was kept for further analyses \cite{Li2013}.



\Genetics2level{Population structure}
Principal component analyses (PCA) were run on scaled and centered predictor 
matrices using the \emph{pca} function from the R-package \textit{LEA} 
\cite{Frichot2015}.
Estimates of ancestry coefficients were estimated based on sparse nonnegative
matrix factorization algorithms \cite{Frichot2014} using the \emph{snmf} 
function in \textit{LEA}.
Algorithms were run for $K=2, \dots 5$ putative ancestral gene pools with 25
repetitions for each value of $K$. 
For each $K$, only the repetition with the lowest cross-entropy was kept for 
further analyses.



\Genetics2level{Core sampling}
In order to evaluate the influence of the genetic constitution of the set of
genotypes that has only data on one out of two predictors, we generated core 
samples using the R-package \textit{corehunter} \cite{Thachuk2009}.
These core samples were based on the 149 maize inbred lines from the diversity
panel which had gene expression as well as genomic information.
The size of the core set was varied in increments of ten percentage points and 
ranged from 10\% to 90\% of all 149 inbred lines.
For assembling the various core sets, we first computed the Modifed Rogers (MR)
distance between each pair of genotypes and then selected a predefined subset
of genotypes that display maximum average distance between each other.


\Genetics2level{Prediction}
\subsection{Kernels}
BLUP models are computationally attractive when the number of features $p$
exceeds the number of genotypes $n$ and are equivalent to a selection index
when fixed effects have been accomodated by the dependent variable 
\cite{Mrode2014}.
Depending on the data set, up to three predictors were available for agronomic
trait predictions, namely pedigree data (P), genomic data (G) and
transcriptompic data (T).
The corresponding feature matrix of the $l$-th group of inbred lines and the 
$o$-th predictor ($\mathbf{W}_{lo}$) has dimensions $n \times p$ where $n$
pertains to the number of genotypes in the $l$-th group and $p$ pertains to 
the number of features.
For the diversity panel maize lines, $n$ corresponds to the number of genotypes
whereas, in the case of the hybrid data, $n$ corresponds to the number of
parent lines in the corresponding heterotic group.
In generating kernels for each predictor, all features in $\mathbf{W}_{lo}$ were
centered and standardized to unit variance.
Then, each kernel can be defined as

\begin{equation} \label{eq:GenomicRelationship}
  \mathbf{K}_{lo} = \frac{1}{W_{lo}} \mathbf{W}_{lo} \mathbf{W}_{lo}^{\top},
\end{equation}

where $W_{lo}$ denotes the number of features \cite{VanRaden2008}.
In the case of pedigree data (P), coancestry coefficients were used directly
for $\mathbf{K}_{lo}$.





\subsection{Universal breeding value model}
The universal model for breeding values was:

\begin{equation} \label{eq:KBLUPModel}
  \mathbf{y} = \mu + 
  \sum_{l=1}^{L} \mathbf{Z}_{lo} \mathbf{g}_{lo} +
  \mathbf{\epsilon},
\end{equation}


where $\mathbf{y}$ is the vector of observed inbred line or hybrid performance,
respectively, $\mu$ is the fixed model intercept, $\mathbf{Z}_{lo}$ is the 
corresponding design matrix associating the random genotype effects 
(\textit{i.e.}, breeding values) of the lines in the $l$-th inbred line group 
and the $o$-th predictor type with $\mathbf{y}$.
Note that, in the case of hybrid data, $\mathbf{Z}_{lo}$ associates the random
general combining ability effects in the parental inbred lines (either Dent or
Flint, as expressed by $l$) with the vector of observed hybrid performance
$\mathbf{y}$.
The random effects ($\mathbf{g}_{lo}$) have expectation zero and variance equal 
to $\mathbf{K}_{lo} \sigma^{2}_{{g}^{lo}}$ and $\mathbf{I} \sigma^2_{\epsilon}$ 
for the residual error.
All prediction models were implemented using the R-package \textit{BGLR}
\cite{Perez2014}. 





\subsection{Single step prediction}
The breeding values in a prediction model are in general given by:
\begin{equation} \label{eq:mrnaebv}
	\mathbf{\tilde{g}} = \mathbf{W}\boldsymbol{\hat{\alpha}},
\end{equation}

where $\boldsymbol{\hat{\alpha}}$ is the solution to a ridge regression model,
which is equivalent to the BLUP model.

Consider now the situation in which one predictor is complete in the sense that
it contains information on the whole population whereas a second predictor is
incomplete and covers only a subset of the population.
We can take a similar route as in \citeNP{Fernando2014} and impute covariates 
of the incomplete predictor by using covariates of the complete predictor.
Let the subscript $1$ denote individuals covered only by the complete
predictor whereas individuals covered by both, the complete and the
incomplete predictor, are indicated by the subscript $2$.
Further, the covariates in $\mathbf{W}$ are centered.
The vector $\mathbf{g}_1$ can be written as the sum of the conditional 
expectation given $\mathbf{g}_2$ and a residual error term:
\begin{align} \label{eq:mrna1}
	\mathbf{g_1} &= E(\mathbf{g}_1|\mathbf{g}_2) + \boldsymbol{\epsilon} \\
	&= \mathbf{K_{12}}\mathbf{K_{22}}^{-1}\mathbf{W_2}\boldsymbol{\hat{\alpha}} + (\mathbf{g_1} - \mathbf{K_{12}}\mathbf{K_{22}}^{-1}\mathbf{W_2}\boldsymbol{\hat{\alpha}}) \\
	&= \mathbf{\hat{g}}_1 + \boldsymbol{\epsilon}
\end{align}

The covariance matrix of $\boldsymbol{\epsilon}$ is $\mathbf{K}_{11} - \mathbf{K}_{12}\mathbf{K}_{22}^{-1}\mathbf{K}_{21} = (\mathbf{K}^{11})^{-1}$
\cite{Legarra2009}.

The structure of the residual imputation/prediction error is known and can 
therefore be modelled.
The covariates not covered by the incomplete predictor can be predicted using 
the expectation of a multivariate normal random vector given correlated 
observations, which are related to the Best Linear Predictor.

The general model for the single step procedure can then be written as 

\begin{equation} \label{eq:single-step-model}
\mathbf{y} = \mathbf{Xb} + \mathbf{W} \boldsymbol{\alpha} + \mathbf{U}^{Complete} \boldsymbol{\epsilon}^{Complete} + \mathbf{e},
\end{equation}

with

\begin{equation} \label{eq:single-step-submatrices}
\mathbf{X} = 
\begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix},
 \mathbf{W} = 
\begin{bmatrix}
  Z_1\hat{\mathbf{W}_1} \\
  \mathbf{W}_2 
 \end{bmatrix},
 \mathbf{U}^{Complete} = 
\begin{bmatrix}
  Z_1 \\
  0 
 \end{bmatrix}
\end{equation}

Following the notation in \citeNP{Fernando2014} the phenotypes can be untangled 
like this:

\begin{align} \label{eq:entangled-augmented-single-step-model}
\begin{bmatrix}
  y_1 \\
  y_2 
 \end{bmatrix}
& =
 \begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix}
 \boldsymbol{\beta} + 
 \begin{bmatrix}
  Z_1 & 0 \\
  0 & Z_2 
 \end{bmatrix}
\begin{bmatrix}
  g_1 \\
  g_2 
 \end{bmatrix}
  + \mathbf{e} \\
    & = 
 \begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix}
 \boldsymbol{\beta} + 
 \begin{bmatrix}
  Z_1 & 0 \\
  0 & Z_2 
 \end{bmatrix}
\begin{bmatrix}
  \mathbf{K}_{12}\mathbf{K}_{22}^{-1}\mathbf{W}_2\boldsymbol{\alpha} + \boldsymbol{\epsilon}^{Complete}  \\
  \mathbf{W}_2\boldsymbol{\alpha} \\
 \end{bmatrix}
  + \mathbf{e} \\
    & = 
 \begin{bmatrix}
  X_1 \\
  X_2 
 \end{bmatrix}
 \boldsymbol{\beta} + 
 \begin{bmatrix}
  Z_1 & 0 \\
  0 & Z_2 
 \end{bmatrix}
\begin{bmatrix}
  \hat{\mathbf{W}_1}\boldsymbol{\alpha} + \boldsymbol{\epsilon}^{Complete} \\
  \mathbf{W}_2\boldsymbol{\alpha} \\
 \end{bmatrix}
  + \mathbf{e} \\
\end{align}

The breeding values from that model are:

\begin{equation} \label{eq:breeding-values}
\tilde{\mathbf{g}} = 
 \begin{bmatrix}
  \hat{\mathbf{W}_1} \\
  \mathbf{W}_2 
 \end{bmatrix}
 \hat{\boldsymbol{\alpha}}
 + 
 \begin{bmatrix}
  \mathbf{Z}_1 \\
  0
 \end{bmatrix}
 \hat{\boldsymbol{\epsilon}}^{Complete} 
\end{equation}



Our final model looks like this:

\begin{align} \label{eq:final-model}
\mathbf{y} &= \mathbf{Xb} +
\sum_{l=1}^{L} \mathbf{Z}_{l}\mathbf{W}_{l} \boldsymbol{\alpha}_{l} + 
\sum_{l=1}^{L} \mathbf{Z}_{l}\mathbf{U}_{l}^{Complete} \boldsymbol{\epsilon}_{l}^{Complete} +
\mathbf{e}.
\end{align}

For the hybrid data, the obtained GCA effects for the inbred lines represent 
half their breeding values.
The predicted agronomic performance is:

\begin{equation} \label{eq:predicted-performance}
\hat{\mathbf{y}} = \sum_{l=1}^{L} \mathbf{Z}_{l}\tilde{\mathbf{g}}_{l} 
\end{equation}




\subsection{Predictive ability and model validation}
For the validation of our predictions, we employed leave-one-out
cross-validation (LOOCV) routines.
In the case of the diversity panel maize inbred lines, LOOCV was performed by 
using a single genotype as a hold-out sample, which will subsequently be 
predicted, while using all other inbred lines for model training.
This process is repeated until all inbred lines have been used once for testing
and $n - 1$ times for model training.

For the hybrid data, LOOCV was carried out as follows:
Let $D$ and $F$ denote the set of parental inbred lines from the Dent and the 
Flint group, respectively.
Further, let $H \cap [D \times F]$ denote the set of hybrids from crosses 
between $D$ and $F$.
The training set ($H_{TRN}$) for the hybrid to be predicted ($H_{ij}$) --- 
where $i \in D_{TRN}$ and $j \in F_{TRN}$ --- was assembled as 
$H_{TRN} = [H \cap (D_{TRN}^{C} \times F_{TRN}^{C})]$ with
$D_{TRN}^{C} = D \setminus D_{TRN}$ and $F_{TRN}^{C} = F \setminus F_{TRN}$.

We judged the performance of each model by looking at its predictive ability,
which is calculated as $\rho(\mathbf{y}, \mathbf{\hat{y}})$, where 
$\mathbf{\hat{y}}$ is the vector of predicted values from each LOOCV run.
Our confidence in the performance of each model was evalutaed by computing the
coefficient of variation for the predicted values:

\begin{equation} \label{eq:CV}
CV = \sqrt{\frac{\sum_{i = 1}^{n} \sigma^{2}_{\hat{y}(i)} / n}{\sum_{i=1}^{n}\hat{y}_{i} / n}},
\end{equation}

where $n$ denotes the number of genotypes (*i.e.*, inbred lines or hybrids).












\section*{Results}
\Genetics2level{Inbred lines}
At the level of the 149 genotypes for which both, genomic (G) and 
transcriptomic (T) information, were available, the use of T yielded higher
predictive abilities than the use of G for four out of six traits
(Table \ref{table:InbredTable}).
In terms of predictive ability, the absolute advantage of T over G was as large 
as 0.04 as observed for the traits '100grainweight', 'Eardiameter' and 
'Silkingtime'.
Except for 'Kernelwidth', the predictive ability for predictions based on
G and the full set of inbred lines ($n = 211$) was always higher than for the 
reduced set of inbred lines ($n = 149$).
The combination of G and T (\textit{i.e.}, GT) in a single-step approach for the 
prediction of all 211 lines performed at least as well as G alone when T was 
superior over G for the reduced set of 149 lines.
In the case of '100grainweight' and 'Eardiameter', the use of GT yielded a 
slight improvement in predictive ability over G.
Where T was a worse predictor for the reduced set of 149 lines than G,
predictive abilities of GT for the full set of 211 lines were also worse than 
those based on G alone.

% latex table generated in R 3.3.3 by xtable 1.8-2 package
% Fri Mar 24 11:34:47 2017
% Script: ./analysis/maizego_core_fraction_plot_predictive_abilities.R
\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for different core sets from the tropical/subtropical material of the maize diversity panel. As predictors, genomic (G) data, transcriptomic (T) data and their combination (GT) were used. The decimal after the predictor name indicates the size of the genetic material that was covered by inbred lines in the core set.} 
\label{table:InbredTable}
\begin{tabular}{rllllll}
  \toprule
 & 100grainweight & cobweight & Eardiameter & Kernelwidth & Plantheight & Silkingtime \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: FALSE}}\\
G Full & 0.43 (0.073) & 0.37 (0.173) & 0.39 (0.053) & 0.34 (0.045) & 0.45 (0.076) & 0.53 (0.03) \\ 
  GT Full & 0.44 (0.075) & 0.33 (0.173) & 0.4 (0.054) & 0.34 (0.043) & 0.4 (0.071) & 0.53 (0.028) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: TRUE}}\\
G 1.0 & 0.36 (0.077) & 0.21 (0.168) & 0.3 (0.053) & 0.34 (0.047) & 0.41 (0.078) & 0.45 (0.03) \\ 
  T 1.0 & 0.4 (0.075) & 0.15 (0.157) & 0.37 (0.051) & 0.36 (0.042) & 0.32 (0.069) & 0.49 (0.028) \\ 
  GT 0.1 & 0.29 (0.076) & 0.21 (0.166) & 0.23 (0.052) & 0.32 (0.047) & 0.4 (0.074) & 0.39 (0.03) \\ 
  GT 0.2 & 0.35 (0.072) & 0.23 (0.164) & 0.26 (0.052) & 0.36 (0.047) & 0.38 (0.072) & 0.44 (0.027) \\ 
  GT 0.3 & 0.36 (0.071) & 0.2 (0.157) & 0.28 (0.05) & 0.37 (0.044) & 0.42 (0.074) & 0.5 (0.028) \\ 
  GT 0.4 & 0.39 (0.07) & 0.2 (0.156) & 0.33 (0.049) & 0.35 (0.043) & 0.38 (0.073) & 0.52 (0.028) \\ 
  GT 0.5 & 0.41 (0.067) & 0.16 (0.155) & 0.31 (0.049) & 0.4 (0.041) & 0.44 (0.075) & 0.49 (0.028) \\ 
  GT 0.6 & 0.4 (0.074) & 0.13 (0.156) & 0.27 (0.05) & 0.37 (0.042) & 0.42 (0.075) & 0.55 (0.027) \\ 
  GT 0.7 & 0.37 (0.073) & 0.1 (0.156) & 0.31 (0.051) & 0.37 (0.042) & 0.33 (0.075) & 0.52 (0.028) \\ 
  GT 0.8 & 0.4 (0.077) & 0.21 (0.158) & 0.36 (0.051) & 0.41 (0.046) & 0.29 (0.074) & 0.49 (0.029) \\ 
  GT 0.9 & 0.39 (0.079) & 0.21 (0.169) & 0.38 (0.056) & 0.41 (0.046) & 0.29 (0.075) & 0.49 (0.031) \\ 
  GT A1 & 0.32 (0.079) & 0.03 (0.155) & 0.28 (0.053) & 0.32 (0.045) & 0.31 (0.073) & 0.46 (0.03) \\ 
  GT A2 & 0.4 (0.081) & 0.18 (0.164) & 0.32 (0.055) & 0.33 (0.046) & 0.36 (0.073) & 0.47 (0.03) \\ 
  GT A3 & 0.3 (0.071) & 0.1 (0.164) & 0.34 (0.054) & 0.41 (0.041) & 0.33 (0.072) & 0.41 (0.029) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}





\Genetics2level{Hybrids}
The maize hybrid data set contains information on six agronomic traits and
three predictors, namely pedigree data (P), genomic information (G), 
transcriptomic information (T), and combinations thereof.
Pedigree records and genomic information were available for the 245 parent 
lines of all 1,521 hybrids and we denote these data as the 'Full' set whereas
transcriptomic data were available for 103 parent lines of 685 hybrids, which
we denote as 'Core'.
For dry matter yield (DMY), T was the best predictor in the core set of 685 
hybrids with a predictive ability that was 0.08 points higher than for genomic 
data, as the second best predictor for this trait 
(Table \ref{table:HybridTable}).
In the core set, P had the highest predictive ability for starch ('STA').
For the other four traits, G had the best predictive abilities.
In the full set of 1,521 hybrids, predictive abilities obtained with G rose 
markedly for all traits with the exception of 'FAT'.
The combination of genomic and transriptomic information (GT) in a single-step
model yielded a small improvement of G alone for the trait 'DMY'.
All traits, where T was not superior to G in the core set, did not benefit from 
the single-step approach when compared to genomic information alone.
Predictive abilities resulting from combinations of pedigree with genomic (PG)
information were better than P alone for all traits and PT improved upon P in
the core set for four traits.

% latex table generated in R 3.3.3 by xtable 1.8-2 package
% Fri Mar 24 12:24:30 2017
% Script: ./analysis/uhoh_plot_predictive_abilities.R
\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for the set of maize hybrids. As predictors, pedigree (P), genomic (G), transcriptomic (T) data and combinations thereof we used.} 
\label{table:HybridTable}
\begin{tabular}{lllllll}
  \toprule
Predictor & DMY & DMC & FAT & PRO & STA & SUG \\ 
  \midrule
P Full & 0.54 (0.049) & 0.48 (0.052) & 0.21 (0.037) & 0.6 (0.019) & 0.35 (0.043) & 0.42 (0.117) \\ 
  G Full & 0.74 (0.052) & 0.63 (0.052) & 0.43 (0.036) & 0.65 (0.022) & 0.46 (0.046) & 0.47 (0.126) \\ 
  PG Full & 0.71 (0.054) & 0.57 (0.055) & 0.35 (0.038) & 0.62 (0.023) & 0.39 (0.048) & 0.42 (0.13) \\ 
  PT Full & 0.73 (0.056) & 0.53 (0.059) & 0.35 (0.04) & 0.61 (0.025) & 0.31 (0.05) & 0.37 (0.14) \\ 
  GT Full & 0.75 (0.053) & 0.61 (0.055) & 0.43 (0.037) & 0.65 (0.024) & 0.44 (0.048) & 0.44 (0.134) \\ 
  P Core & 0.46 (0.047) & 0.53 (0.05) & 0.23 (0.037) & 0.46 (0.023) & 0.38 (0.039) & 0.34 (0.132) \\ 
  G Core & 0.67 (0.053) & 0.47 (0.055) & 0.43 (0.038) & 0.48 (0.027) & 0.36 (0.046) & 0.35 (0.132) \\ 
  T Core & 0.75 (0.055) & 0.34 (0.062) & 0.4 (0.041) & 0.47 (0.029) & 0.16 (0.05) & 0.27 (0.145) \\ 
   \bottomrule
\end{tabular}
\end{table}






\Genetics2level{Coverage of the genetic space}
Single-step approaches are based on the use of one complete predictor that
covers all genotypes and one incomplete predictor that covers only a subset of
all available genotypes.
Here, we explore the impact of which data are covered by the incomplete 
predictor on the predictive abilities obtained via single-step approaches by
focusing exclusively on the 149 inbred lines from the maize diversity panel
that are covered by both, genomic and transcriptomic information.
First, we assmbled nine different core sets, which we defined as subsets
of genotypes that maximize the average genetic distance between genotypes of
the full set.
The nine core sets differ in that each covers a different number of genotypes,
raning from only 10\% to to 90\% of all genotypes covered by the incomplete
predictor (Fig. \ref{fig:CoreSetPCA}).


\begin{figure}[H]
\centering
\includegraphics{./tables_figures/maizego_core_sampling_pca.pdf}
\caption{
  Coverage of the genetic space by nine different core sets. 
  Genotypes are distributed according to their scores based on the first two
  principal components (PC).
  The total genetic space was spanned by 149 genotypes from the maize diversity 
  panel for which both, genomic and transcriptomic data, were available.
  Genotypes that were members of the core set (orange triangles) were declared 
  as being covered by both predictors whereas members of the complement (blue,
  filled circles) were declared as only being covered by genomic information.
  Assignments of genotypes to the core set was done by selecting a subset of
  genotypes that maximized the average genetic distance in the core set based 
  on the Modified Rogers distance between all genotypes.
  Core set sizes ranged from 10\% to 90\% of the 149 available genotypes and
  are given as decimals above each facet of the plot.
}
\label{fig:CoreSetPCA}
\end{figure}


Predictive abilities obtained with the different core sets varied widely within
traits (Fig. \ref{fig:CoreSetInfluence}, Table \ref{table:InbredTable}).
For the traits 'cobweight' and 'Plantheight', which were better predicted via G
than via T in the reduced set of lines, predictive abilities were higher when
only a small number of genotypes was included in the core set
For most other traits, an initial increase in predictive abilities when moving
from a core set size of 0.1 to 0.5/0.6 was observed, followed by a drop around
0.6/0.7 and a subsequent hike up to a core set size of 0.9.



\begin{figure}[H]
\centering
\includegraphics{./tables_figures/core_fraction_predictive_ability_trend.pdf}
\caption{
Influence of core set composition on predictive abilities.
Predictive abilities are based on single step predition with genomic data as the 
complete predictor and transcriptomic data as the incomplete predictor.
Results are shown for six agronomic traits and nine core sets of the maize 
diversity panel.
}
\label{fig:CoreSetInfluence}
\end{figure}


As an alternative to exploring the impact of genotypes sampled based on the
maximization of genetic distance within the core set, we wanted to explore a
more extreme scenario where we select core sets based on population structure.
Hereto, we ran an admixture analysis on the same set of 149 inbred lines and
display the estimated ancestry coefficients for the genotypes assuming $K=3$
ancestral populations (Fig. \ref{fig:PopStructure}).

\begin{figure}[H]
\centering
\includegraphics{./tables_figures/maizego_admixture_k3_pca.pdf}
\caption{
  Population structure of 149 maize inbred lines from the maize diversity panel
  covered by genomic and transcriptomic data.
  (\textbf{A}) Ancestry coefficients from an admixture analysis when 
  considering $K=3$ ancestral populations.
  (\textbf{B}) Principal component analysis with genotypes colored according to 
  the most likely ancestral population (\textit{i.e.}, share $\geq 0.5$) as 
  indicated by the admixture analysis.
  Genotypes for which none of the three ancestry coefficients reached a value
  of 0.5 were classified as "mixed".
}
\label{fig:PopStructure}
\end{figure}


Of the 149 genotypes, 17 belonged to more than 50\% to the first ancestral
population (A1), 109 belonged to more than 50\% to the second ancestral
population (A2) and 19 belonged to more than 50\% to the third ancestral
population while four genotypes could not be unequivocally assigned to either
ancestral population (mixed).
To adjust the size of the set of genotypes largely belonging to A2 to the other
two sets, we selected 19 genotypes from A2 using previously described core
sampling.
For single-step prediction, each of the three genotype sets were then declared
as having genomic and transcriptomic information whereas their complements were
covered by transcriptomic data only.
Genotypes from A1 and A3, which were at the far ends in the PCA plot (Fig.
\ref{fig:PopStructure}B), were estimated with less precision than genotypes
from the central cluster of genotypes belonging largely to the A2 population
(Fig. \ref{fig:PopStructure}A, Table \ref{table:InbredTable}).


A principal component analysis, carried out separately on the two parental
heterotic groups of the maize hybrids, showed that the entire genetic space in
each group was well covered by genotypes that had both, transcriptomic and
genomic data (Fig. \ref{fig:UHOH-PCA}).

\begin{figure}[H]
\centering
\includegraphics{./tables_figures/uhoh_dent_flint_pca.pdf}
\caption{
  Principal component (PC) analysis of the two heterotic groups of the maize 
  hybrids.
  All inbred lines, which are covered by genomic and transcriptomic features
  are depicted as orange triangles whereas genotypes for which only genomic
  information are available are depicted as blue dots.
}
\label{fig:UHOH-PCA}
\end{figure}













\section*{Discussion}
\Genetics2level{Advantages of marker effect models}
% Describe the many hacks that have been developed to tackle problems with BLUP
% models.

$\dots$

\Genetics2level{Superiority of single step models}


\Genetics2level{Coverage of the genetic space}
Animal populations are so large that it does not truly matter, which animals are 
being genotyped, but in plant breeding the situation is different and we would 
be interested in determining which individuals should be complemented with 
information on another predictor.


Prediction of phenotypic performance is particularly crucial in hybrid breeding
programs because the number of potential selection candidates is $n^{2}$ where
$n$ corresponds to the number of parent lines.

Given recent results on the utility of downstream 'omics' predictors in a study
on hybrid prediction in a European hybrid maize panel \cite{Westhues2017}, our 
objectives were i) to explore the benefits of a single step procedure for the 
same material and ii) to examine the influence on the composition of the set of
genotypes that is covered by more than one predictor on the success of single
step prediction.




\section*{Conclusions}


$\dots$




\section{Acknowledgments} 
This project was funded by the German Federal Ministry of Education and 
Research (BMBF) within the projects OPTIMAL (FKZ: 0315958B,0315958F),
SYNBREED (FKZ: 0315528D) and by the German Research Foundation 
(DFG, Grants No. ME 2260/5-1 and SCHO 764/6-1).
The authors acknowledge support by the state of Baden-W{\"u}rttemberg through 
bwHPC.
Financial support for M.W. was provided by the Fiat Panis foundation, Ulm,
Germany.



\nolinenumbers
% Bibliography
\bibliography{library}
\bibliographystyle{mychicago}
\end{document}
