# Aguilar 2009 JDairySci
-   "A typical evaluation erquires 1) traditional evaluation with an animal model,
    2) extraction of pseudo-observations such as deregressed evaluations or
    daughter deviations (DD), 3) estimation of genomic effects for genotypes
    animals usually using simple sire models, and possibly 4) combining the genomic
    index with traditional parent averages (PA) and EBV (Hayes 2009, VanRaden
    2009b)."

-   "Genomic evaluations are more accurate than PA and approach the accuracy of
  evaluations for progeny-tested bulls, but they also seem inflated (VanRaden
  2009a). Although their inflation is lower than that of current PA, the
  potentially great utilization of top genomically evaluated young sires
  increases the improtance of high accuracy and minimum bias. Inflation of
  genetic evalutions by genomic information causes top young bulls to have an
  unfair advantage over older progeny-tested bulls. Some of the problems with
  genomic evaluations may be caused by incorrect parameters and strong
  assumptions used in multiple-step procedures. However, effects of those
  parameters and assumptions are extremely difficult to verify, particularly in
  the presence of selection."

-   "A more serious problem is when pseudo-observations are poorly defined or
  of poor quality (e.g. for animals with small progeny numbers), which is often
  the case for monogastric species and for beef cattle."

-   "The objective of this study was to use a single-step procedure for genomic
    evaluation in a national evaluation setting and compare its performance to
    a multi-step procedure."


# Ashraf 2016 CropSci
-   "The objective of this study was to evaluate single-step GP in a wheat
  program. We compared the performance of pedigree-based, marker-based and
  single-step models (ABLUP, GBLUP and HBLUP, respectively). Data consisted of
  1,176 genotypes (via genotyping-by-sequencing) and 11,131 nongentyped wheat
  lines."

-   In a dairy cattle study, Su et al. (2012) reported that this single-step
  approach gave a more accurate GEBV than a multistep procedure."

-   "The main aim of this study was to investigate the impact on prediction
  accuracy when some wheat lines are not genotyped. In additiona, we compared
  the performance of different prediction models and determined the impact of
  different weights for the pedigree relationship on prediction accuracy."

-   "The single-step method (Christensen et al. 2012, Legarra et al. 2009)
  includes information from genotyped and nongentoyped individuals by blending
  the genomic relationship matrix $\mathbf{G}$ and the pedigree-based matrix
  $\mathbf{A}$. The HBLUP method uses the same linear model as the GBLUP
  methods with the exception that $u \texttilde N(0,
  \mathbf{H}\sigma^{2}_{g}$), where $\mathbf{H}$ is a combined relationship
  matrix (Christensen et al. 2012, Legarra et al. 2009)."

-   "We used $w = 0.05$ to place the majority of the weight on the genomic
  relationship matrix as this matrix should be able to track Mendelian
  segregation and we therefore assumed that it would be a more accurate
  estimate of the true relationship. A similar $w$ was also tested in
  Christensen et al. (2012)."

-   Optimized the weight of genomic and pedigree information in blending the
  two matrices by fitting them simultaneously as random effects in a mixed 
  model.

-   "The polygenic effect weight, $w$, optimized in the HBLUP-weight model,
  outperformed the fixed weight used in HBLUP for all traits.

-   "First, HBLUP accuracy was greater than GBLUP accuracy in all caes. Similar
  results were reported by Gao et al. (2012) in Nordic Holstein population. The
  likely reason for this superiority is that HBLUP was able to use much more
  phenotypic information than GBLUP. From Table 1, we can see than on average
  over all traits, there were close to 10,000 observations on nongenotyped
  lines. These observations were not available to GBLUP, for which there were
  generally < 7000 observations across six environments."

-   "Optimzing the weight $w$ for combining the pedigree and the genomic
  relationship matrices in scenario 2 (termed HBLUP-weight here) gave higher
  accuracies than fixed weight HBLUP in all traits. The optimal weighting
  factors differed between the traits."

-   "One interpretation of the different weights between traits is that this
  parameter is affected by the fraction of the variance caused by low minor
  allele frequency (MAF) loci whose effects are poorly captured by markers. For
  traits with a high fraction of the variance generated by such low MAF causal
  loci, the weight would be high on the pedigree component."


# Christensen and Lund 2010, GSE
-   "The extension of the genomic relationship matrix to non-gneotyped animals
  together with the addition of the polygenic effect provide a natural one-step
  procedure to blend the information from relatives and the genomic information
  into a combined genomically enhanced breeding value (GEBV)."


# Christensen et al. 2012, Animal
-   "The initial methods for incorporating genomic information into genetic
  evaluation were multi-step. These methods consist of (1) constructing a
  response variable for the genotyped animals that summarises the phenotypic
  information, (2) a genomic prediction method that associates the responsive
  variable to the marker information and (3) blending genomic predictions with
  parent average estimated breeding values (EBV)."

-   "Single-step methods provide predictions for all animals in the pedigree."

-   "Furthermore, the single-step methods include information about Mendelian
  inheritance through markers, thereby providing a more accurate modelling, and
  we therefore conjecture that for non-gneotyped animals single-step methods
  also produce more accurate predictions than the pedigree-based method."

-   "The single-step method in theory requires that the genomic relationship
  matrix is constructed using the allele frequencies in the founder population
  of the pedigree, but these are not feasible to compute with sufficient
  accuracy in practice since founder animals are ussually not genotyped. Two
  obvious alternatives are to use either observed allele frequencies or assume
  that allele frequencies are equal to 0.5 instead. Forni et al. (2011) and
  Chen et al. (2011) concluded that using observed allele frequencies resulted
  in the most accurate predictions."

-   "The specific aim of this study on DG and FCR in Danish Duroc pigs were to
  demonstrate that (1) single-step methods are at least as accurate as GBLUP
  for prediction of genotyped animals, and that both outperform the
  pedigree-based method in this respect, (2) single-step methods provide more
  accurate predictions than the pedigree-based method for non-genotyped
  animals, (3) adjusted single-step method produces more accureate predictions
  and behave better than the original single-step method, and 94) bivariate
  models improve accuracy of predictions for FCR, no matter which of the
  methods are used."

-   "Here we combined these two ideas and adjusted $\mathbf{G}$ to
  $\mathbf{G}_{a} = \beta \mathbf{G} + \alpha$, where $\beta$ and $\alpha$
  solved the system of equations $\text{Avg}(diag(\mathbf{G}))\beta + \alpha =
  \text{Avg}(diag(\mathbf{A}_{11}))$."

-   "The comparison of the different methods was therefore based on these
  validation correlations, and differences were assessed using the
  Hotelling-Williams *t*-test at confidence level 5%."

-   Contrary to what the authors claim, the differences in $r$ between regular
  single step prediction and single step predictions with adjustments to the
  genomic relationship matrix are minor (third digit after the comma, see Table
  2).

-   "The results show that the three methods with marker information (two
  single-step methods and GBLUP) produce more accurate predictions of genotyped
  animals than the pedigree-based method. In addition, single-step methods
  produce more accurate predictions for non-genotyped animals."

-   "In this paper, differences in validation correlation between methods are
  assessed with the Hotelling-Williams *t*-test. Most previous paper on genomic
  evaluation have not made any assessment of statistical significance of
  differences in accuracy between methods. The likely reason for this is that
  confidence intervals for accuracies of different methods are often highly
  overlapping, which falsely gives the impression of statistical
  non-significance. The Hotelling-Williams *t*-test takes into account the
  strong correlation of predictions from different methods, and is therefore a
  much more powerful test. The test is known in psychology (Revelle, 2010), but
  does not seem to be much known in other scientific fields."


# Christensen 2012, GSE
-   "In this paper, I propose a new method to address this issue by 1)
  adjusting the pedigree-based relationship matrix to be compatible with the
  marker-based relationship matrix instead of the reverse and 2) extending the
  single-step genetic evaluation using a joint likelihood of observed
  phenotypes and observed marker genotypes."

-   "A single-step method extends the usual pedigree-based method by replacing
  the additive relationship matrix constructed from pedigree by an additive
  relationship matrix that combines the marker-based relationship matrix for
  genotyped animals with the pedigree-based relationship matrix."

-   "An issue with a single-step method is compatability between the
  marker-based relationship matrix and the pedigree-based relationship matrix.
  To handle this problem, it is necessary to determine which allele frequencies
  should be used in the marker-based relationship matrix and to adjust this
  matrix to the pedigree-based relationship matrix. In theory, one should use
  the allele frequencies in the founder population of the pedigree (base
  animals) for the marker-based and pedigree-based relationships to be
  compatible, but these allele frequencies are rarely availalbe in practice
  since base animals are not genotyped."

-   "This results in a single-step method in which the marker-based
  relationship matrix is constructed assuming all allele frequencies are equal
  to 0.5 and the pedigree-based relationship matrix is constructed using the
  unusual assumption that animals in the base population are related and inbred
  with a relationship coefficient $\gamma$ and an inbreeding coefficient
  $\gamma / 2$."


# Christensen et al. 2014, GSE
-   "Production systems based on crossbreeding are predominant in pig and
  chicken breeding and take advantage of the increased performance of crossbred
  animals compared to purebred animals."

-   "Genomic selection of purebreds for crossbred performance was proposed by
  Ibanez-Escriche et al. (2009) that used phenotypes on crossbreds only, and a
  genomic model with a breed of origin specific allele substitution effects."

-   "A single-step method for genomic evaluation of both purebred and crossbred
  performances has not been developed, yet."

-   "A partial relationship matrix is a relationship matrix that describes the
  relationships only according to genetic origin."


# Fernando, Dekkers and Garrick 2014, GSE
-   "Phenotypic values or deregressed estimated breeding values (EBV) on these
  genotyped individuals are used to estimate the effects of the SNPs using
  Bayesian multiple-regression models in which the marker effects are treated
  as random. We refer to such models as marker effect models (MEM)."

-   "Fernando [7] showed that the MEM and BVM provide equivalent predictions of
  BV and that predictions could be more easily obtained from the MEM than from
  BVM because at the time of their study the number of markers was much smaller
  than the number of genotyped animals."

-   "Thus, their method is called single-step BLUP (SS-BLUP), and is expected
  to yield unbiased predictions under multivariate normality, even in
  populations that are undergoing selection and non-random mating. This is
  important, because genotypes are usually collected only on superior animals
  and this can lead to biased evaluations. A properly calculated BLUP
  evaluation has been shown to be free of this selection bias [15-21]."

-   "Conceptually, the remaining relationship coefficients constructed from
  pedigree are modified to provide consistency with $\mathbf{G}$ [11]."

MANY MORE NOTES


# Fernando et al. (2016), GSE
-   "The objective of this paper is to present computational strategies for
  whole-genome analyses based on the SS-MEM that avoid storing large, dense
  blocks of the MME that involve imputed genotypes. First, we will show this
  for the MME given in [10]. Second, we will present what we refer to as a
  hybrid type model (HM) that uses a MEM for the animals with marker genotypes
  and a BVM for animals without genotypes."

-   "$\mathbf{M}_{g}$ is the matrix of centered marker covariates for animals
  with genotypes, $\mathbf{M}_{n} =
  \mathbf{A}_{ng}\mathbf{A}_{gg}^{-1}\mathbf{M}_{g}$, is the matrix of imputed
  marker covariates for animals with missing genotypes, $\mathbf{\alpha}$ is
  the vector of random marker effects, $\mathbf{\epsilon}$ is the vector of
  imputation residuals with null means and covariance matrix proportional to
  the inverse of $\mathbf{A}^{nn}$, the sub-matrix corresponding to animals
  with missing genotypes in the inverse of the matrix $\mathbf{A}$ of
  pedigree-based additive relationships, and $\mathbf{e}$ is a vector of
  residuals. The matrix of imputed genotypes can be more efficiently computed
  by solving the sparse system of equations [10]:
  $\mathbf{A}^{nn}\mathbf{M}_{n} = - \mathbf{A}^{ng}\mathbf{M}_{g}$."

-   "The large dense matrix $\mathbf{M}_{n}$ appears in the MEM given by Model
  (1). This is avoided here by using a BVM for animals with missing genotypes
  rather than expressing their breeding values as the sum of the effects of
  their imputed marker gneotypes plus their separate imputation residuals. The
  advantages of the MEM such as allowing for alternative priors for marker
  effects are retained by still fitting a MEM but only for animals with
  genotypes."

-   Correlations between estimates of breeding values from the hybrid model and
  the regular model were 1.0 for non-genotyped animals and over 0.99 for
  genotyped animals.


