# Results


\begin{figure}[H]
\centering
\includegraphics{./tabs_figs/core_fraction_predictive_ability_trend.pdf}
\caption{
Influence of core set composition on predictive abilities.
Predictive abilities are based on single step predition with genomic data as the 
complete predictor and transcriptomic data as the incomplete predictor.
Results are shown for six agronomic traits and nine core sets of the maize 
diversity panel.
}
\label{fig:CoreSetInfluence}
\end{figure}

<!--
% latex table generated in R 3.3.3 by xtable 1.8-2 package
% Wed Mar 15 10:35:35 2017
# Script: ./analysis/maizego_plot_predictive_abilities.R
-->
\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for the set of 211 tropical/subtropical lines from the maize diversity panel.} 
\begin{tabular}{rllllll}
  \toprule
 & 100grainweight & cobweight & Eardiameter & Kernelwidth & Plantheight & Silkingtime \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: no}}\\
snp-mrna & 0.44 (0.0748) & 0.33 (0.1727) & 0.4 (0.0543) & 0.34 (0.0425) & 0.4 (0.0707) & 0.53 (0.0285) \\ 
  snp-none & 0.43 (0.0733) & 0.37 (0.1734) & 0.39 (0.0535) & 0.34 (0.0455) & 0.45 (0.0755) & 0.53 (0.0297) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: yes}}\\
mrna-none & 0.4 (0.0747) & 0.15 (0.1568) & 0.37 (0.0509) & 0.36 (0.0423) & 0.32 (0.0692) & 0.49 (0.0279) \\ 
  snp-none1 & 0.36 (0.0773) & 0.21 (0.1678) & 0.3 (0.0527) & 0.34 (0.0472) & 0.41 (0.078) & 0.45 (0.0302) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}



<!--
% latex table generated in R 3.3.3 by xtable 1.8-2 package
% Wed Mar 15 11:11:37 2017
# Script: ./analysis/uhoh_plot_predictive_abilities.R
-->
\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for the set of maize hybrids} 
\begin{tabular}{rllllll}
  \toprule
 & DMY & DMC & FAT & PRO & STA & SUG \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: FALSE}}\\
ped-none & 0.54 (0.0493) & 0.48 (0.0521) & 0.21 (0.0365) & 0.6 (0.0187) & 0.35 (0.0427) & 0.42 (0.117) \\ 
  snp-none & 0.74 (0.0517) & 0.63 (0.0521) & 0.43 (0.0356) & 0.65 (0.0223) & 0.46 (0.0458) & 0.47 (0.1261) \\ 
  ped-snp & 0.71 (0.054) & 0.57 (0.0545) & 0.35 (0.0375) & 0.62 (0.0235) & 0.39 (0.0478) & 0.42 (0.1298) \\ 
  ped-mrna & 0.73 (0.056) & 0.53 (0.0594) & 0.35 (0.0398) & 0.61 (0.0254) & 0.31 (0.0504) & 0.37 (0.1396) \\ 
  snp-mrna & 0.75 (0.0533) & 0.61 (0.0553) & 0.43 (0.0374) & 0.65 (0.0239) & 0.44 (0.048) & 0.44 (0.1342) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: TRUE}}\\
mrna-none & 0.75 (0.0552) & 0.34 (0.0619) & 0.4 (0.0412) & 0.47 (0.0293) & 0.16 (0.0496) & 0.27 (0.1447) \\ 
  ped-none1 & 0.46 (0.0469) & 0.53 (0.05) & 0.23 (0.0368) & 0.46 (0.0234) & 0.38 (0.0391) & 0.34 (0.1316) \\ 
  snp-none1 & 0.67 (0.0534) & 0.47 (0.0552) & 0.43 (0.0384) & 0.48 (0.027) & 0.36 (0.0457) & 0.35 (0.132) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}




<!--
% latex table generated in R 3.3.3 by xtable 1.8-2 package
% Wed Mar 15 11:36:08 2017
# Script: ./analysis/maizego_core_fraction_plot_predictive_abilities.R
-->
\begin{table}[ht]
\centering
\caption{Predictive abilities and corresponding coefficients of variation for different core sets from the tropical/subtropical material of the maize diversity panel.} 
\begin{tabular}{rllllll}
  \toprule
 & 100grainweight & cobweight & Eardiameter & Kernelwidth & Plantheight & Silkingtime \\ 
  \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: FALSE}}\\
snp-mrna & 0.44 (0.0748) & 0.33 (0.1727) & 0.4 (0.0543) & 0.34 (0.0425) & 0.4 (0.0707) & 0.53 (0.0285) \\ 
  snp-none & 0.43 (0.0733) & 0.37 (0.1734) & 0.39 (0.0535) & 0.34 (0.0455) & 0.45 (0.0755) & 0.53 (0.0297) \\ 
   \midrule
\multicolumn{6}{l}{{\bfseries Only incomplete predictor: TRUE}}\\
mrna-none\_1.0 & 0.4 (0.0747) & 0.15 (0.1568) & 0.37 (0.0509) & 0.36 (0.0423) & 0.32 (0.0692) & 0.49 (0.0279) \\ 
  snp-mrna\_0.3 & 0.36 (0.0711) & 0.2 (0.1566) & 0.28 (0.0503) & 0.37 (0.044) & 0.42 (0.0736) & 0.5 (0.028) \\ 
  snp-mrna\_0.4 & 0.39 (0.0697) & 0.2 (0.1558) & 0.33 (0.0486) & 0.35 (0.0428) & 0.38 (0.0727) & 0.52 (0.028) \\ 
  snp-mrna\_0.5 & 0.41 (0.0671) & 0.16 (0.1549) & 0.31 (0.0486) & 0.4 (0.0414) & 0.44 (0.0748) & 0.49 (0.0278) \\ 
  snp-mrna\_0.6 & 0.4 (0.0739) & 0.13 (0.1563) & 0.27 (0.0497) & 0.37 (0.0417) & 0.42 (0.075) & 0.55 (0.0272) \\ 
  snp-mrna\_0.7 & 0.37 (0.0734) & 0.1 (0.1555) & 0.31 (0.0508) & 0.37 (0.0422) & 0.33 (0.0755) & 0.52 (0.028) \\ 
  snp-mrna\_0.8 & 0.4 (0.0774) & 0.21 (0.1581) & 0.36 (0.0515) & 0.41 (0.0458) & 0.29 (0.0737) & 0.49 (0.0288) \\ 
  snp-mrna\_0.9 & 0.39 (0.0792) & 0.21 (0.1692) & 0.38 (0.0561) & 0.41 (0.0459) & 0.29 (0.075) & 0.49 (0.0311) \\ 
  snp-none\_1.0 & 0.36 (0.0773) & 0.21 (0.1678) & 0.3 (0.0527) & 0.34 (0.0472) & 0.41 (0.078) & 0.45 (0.0302) \\ 
   \bottomrule
\multicolumn{6}{l}{}\\
\end{tabular}
\end{table}